{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nComputing entropy at the full epitope level.\\n'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Computing entropy at the full epitope level.\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/larger_version_modified/aligned_protein_E.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_M.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_N.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF10.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF1a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF1b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF3a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF6.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF7a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF7b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF8.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF9b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_S.fasta']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "amino_acid_files = !ls data/larger_version_modified/*_protein_* # newer code\n",
    "amino_acid_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(amino_acid_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Random aside, making a fasta file for each of the full proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n",
      "ind =  0\n"
     ]
    }
   ],
   "source": [
    "seq_dict = dict()\n",
    "for aa_file in amino_acid_files:\n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    new_aa_file= 'data/larger_version_modified/aligned_protein_'+protein+'.fasta' # newer code\n",
    "    with open(new_aa_file, \"rt\") as handle:\n",
    "        records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "        \n",
    "\n",
    "    for ind, r in enumerate(records): \n",
    "        if r.id == 'Wuhan-Hu-1/2019':\n",
    "            print('ind = ', ind)\n",
    "            ref_seq_ind = ind\n",
    "\n",
    "    ref_seq_new = str(records[ref_seq_ind].seq)\n",
    "    seq_dict[protein] = ref_seq_new\n",
    "seq_dict; \n",
    "\n",
    "### All ind's should be zero by design, because I have added the reference sequence to the beginning of all the \n",
    "### protein files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create the 'base_proteins_with_end_codon_larger.fasta' file\n",
    "with open('base_proteins_with_end_codon_larger.fasta', 'w') as f: \n",
    "    for k, v in seq_dict.items():\n",
    "        f.write('> Protein: '+k+' | Base Sequence: Wuhan-Hu-1/2019\\n')\n",
    "        f.write(v +'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Checking if the proteins have changed from original due to new masking\n",
    "\n",
    "* This was necessary when NextStrain was doing different read calling!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for aa_file in amino_acid_files:\n",
    "#     protein = aa_file.split('_')[-1].split('.')[0]\n",
    "#     old_aa_file = '../PUFFIN/proteins/v1/aligned_aa_'+protein+'.fasta' # Trenton's code\n",
    "# #     old_aa_file = 'PUFFIN/proteins/aligned_aa_'+protein+'.fasta' # new code wrong \n",
    "# #     old_aa_file = '../PUFFIN/proteins/v1/aligned_aa_'+protein+'.fasta' # newer code wrong\n",
    "\n",
    "#     print('protein:', protein)\n",
    "#     with open(old_aa_file, \"rt\") as handle:\n",
    "#         records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "#     # getting rid of the node sequences!\n",
    "#     no_nodes = []\n",
    "#     for r in records: \n",
    "#         if 'NODE_' not in r.id:\n",
    "#             no_nodes.append(r)\n",
    "#     records = no_nodes\n",
    "#     len(records)\n",
    "\n",
    "#     for ind, r in enumerate(records): \n",
    "#             if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#                 print(ind)\n",
    "#                 ref_seq_ind = ind\n",
    "\n",
    "#     ref_seq_old = str(records[ref_seq_ind].seq)[:-1]\n",
    "    \n",
    "#     print('length of old sequence', len(ref_seq_old))\n",
    "    \n",
    "#     new_aa_file= 'PUFFIN/proteins/aligned_protein_'+protein+'.fasta'\n",
    "#     with open(new_aa_file, \"rt\") as handle:\n",
    "#         records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "#     # getting rid of the node sequences!\n",
    "#     no_nodes = []\n",
    "#     for r in records: \n",
    "#         if 'NODE_' not in r.id:\n",
    "#             no_nodes.append(r)\n",
    "#     records = no_nodes\n",
    "#     len(records)\n",
    "\n",
    "#     for ind, r in enumerate(records): \n",
    "#             if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#                 print(ind)\n",
    "#                 ref_seq_ind = ind\n",
    "\n",
    "#     ref_seq_new = str(records[ref_seq_ind].seq)[:-1]\n",
    "    \n",
    "#     if str(records[ref_seq_ind].seq)[-1] == '*':\n",
    "#         print(protein, 'has star at end')\n",
    "#     else: \n",
    "#         print(protein, 'has NO STAR at end')\n",
    "    \n",
    "#     mask = np.asarray(list(ref_seq_old)) != np.asarray(list(ref_seq_new)) \n",
    "    \n",
    "#     print(np.asarray(list(ref_seq_old))[mask])\n",
    "#     print(np.arange(len(ref_seq_old))[mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for aa_file in amino_acid_files:\n",
    "#     protein = aa_file.split('_')[-1].split('.')[0]\n",
    "#     old_aa_file = '../PUFFIN/proteins/v1/aligned_aa_'+protein+'.fasta'\n",
    "\n",
    "#     print('protein:', protein)\n",
    "#     with open(old_aa_file, \"rt\") as handle:\n",
    "#         records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "#     # getting rid of the node sequences!\n",
    "#     no_nodes = []\n",
    "#     for r in records: \n",
    "#         if 'NODE_' not in r.id:\n",
    "#             no_nodes.append(r)\n",
    "#     records = no_nodes\n",
    "#     len(records)\n",
    "\n",
    "#     for ind, r in enumerate(records): \n",
    "#             if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#                 print(ind)\n",
    "#                 ref_seq_ind = ind\n",
    "\n",
    "#     ref_seq_old = str(records[ref_seq_ind].seq)[:-1]\n",
    "    \n",
    "#     print('length of old sequence', len(ref_seq_old))\n",
    "    \n",
    "#     new_aa_file= '../PUFFIN/proteins/v6/aligned_protein_'+protein+'.fasta'\n",
    "#     with open(new_aa_file, \"rt\") as handle:\n",
    "#         records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "#     # getting rid of the node sequences!\n",
    "#     no_nodes = []\n",
    "#     for r in records: \n",
    "#         if 'NODE_' not in r.id:\n",
    "#             no_nodes.append(r)\n",
    "#     records = no_nodes\n",
    "#     len(records)\n",
    "\n",
    "#     for ind, r in enumerate(records): \n",
    "#             if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#                 print(ind)\n",
    "#                 ref_seq_ind = ind\n",
    "\n",
    "#     ref_seq_new = str(records[ref_seq_ind].seq)[:-1]\n",
    "    \n",
    "#     if str(records[ref_seq_ind].seq)[-1] == '*':\n",
    "#         print(protein, 'has star at end')\n",
    "#     else: \n",
    "#         print(protein, 'has NO STAR at end')\n",
    "    \n",
    "#     mask = np.asarray(list(ref_seq_old)) != np.asarray(list(ref_seq_new)) \n",
    "    \n",
    "#     print(np.asarray(list(ref_seq_old))[mask])\n",
    "#     print(np.arange(len(ref_seq_old))[mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # what is the frequency before and after at these different sites? \n",
    "# protein_name = 'ORF1b'\n",
    "# position = 1687\n",
    "# old_aa_file = '../PUFFIN/proteins/v1/aligned_aa_'+protein_name+'.fasta'\n",
    "\n",
    "# print('protein:', protein)\n",
    "# with open(old_aa_file, \"rt\") as handle:\n",
    "#     records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "# # getting rid of the node sequences!\n",
    "# no_nodes = []\n",
    "# for r in records: \n",
    "#     if 'NODE_' not in r.id:\n",
    "#         no_nodes.append(r)\n",
    "# records = no_nodes\n",
    "# len(records)\n",
    "\n",
    "# for ind, r in enumerate(records): \n",
    "#     if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#         print(ind)\n",
    "#         ref_seq_ind = ind\n",
    "# aa_counter={}\n",
    "# ref_seq_old = str(records[ref_seq_ind].seq)[:-1]\n",
    "\n",
    "# for ind, r in enumerate(records): \n",
    "#     try: \n",
    "#         aa_counter[str(r.seq[position])] += 1\n",
    "#     except:\n",
    "#         aa_counter[str(r.seq[position])] =1\n",
    "# aa_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new_aa_file = 'data/v6/aligned_protein_'+protein_name+'.fasta'\n",
    "\n",
    "# print('protein:', protein)\n",
    "# with open(new_aa_file, \"rt\") as handle:\n",
    "#     records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "# # getting rid of the node sequences!\n",
    "# no_nodes = []\n",
    "# for r in records: \n",
    "#     if 'NODE_' not in r.id:\n",
    "#         no_nodes.append(r)\n",
    "# records = no_nodes\n",
    "# len(records)\n",
    "\n",
    "# for ind, r in enumerate(records): \n",
    "#     if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "#         print(ind)\n",
    "#         ref_seq_ind = ind\n",
    "# aa_counter={}\n",
    "# for ind, r in enumerate(records): \n",
    "#     try: \n",
    "#         aa_counter[str(r.seq[position])] += 1\n",
    "#     except:\n",
    "#         aa_counter[str(r.seq[position])] =1\n",
    "# aa_counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# amino_acid_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Removing sequences with Gap, X, * or no M to start (except for ORF1b, doesnt start with a M!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E 7 0.09333333333333334 0\n",
      "M 1 0.0045045045045045045 0\n",
      "N 23 0.05489260143198091 0\n",
      "ORF10 39 1.0263157894736843 3\n",
      "ORF1a 12788 2.9063636363636363 0\n",
      "ORF1b 70 0.025974025974025976 0\n",
      "ORF3a 122 0.44363636363636366 3\n",
      "ORF6 48 0.7868852459016393 3\n",
      "ORF7a 63 0.5206611570247934 7\n",
      "ORF7b 90 2.0930232558139537 7\n",
      "ORF8 611 5.049586776859504 10\n",
      "ORF9b 10 0.10309278350515463 0\n",
      "S 715 0.5616653574234093 0\n"
     ]
    }
   ],
   "source": [
    "# doing a quality check \n",
    "error_seq = np.zeros(len(records))\n",
    "for aa_file in amino_acid_files: \n",
    "    \n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    #print('protein:', protein)\n",
    "    with open(aa_file, \"rt\") as handle:\n",
    "        records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "        \n",
    "    # getting rid of the node sequences!\n",
    "    no_nodes = []\n",
    "    for r in records: \n",
    "        if 'NODE_' not in r.id:\n",
    "            no_nodes.append(r)\n",
    "    records = no_nodes\n",
    "    \n",
    "    num_errs = 0\n",
    "    stop_errs =[]\n",
    "    m_errs = 0\n",
    "    \n",
    "    #records = clean_records\n",
    "    \n",
    "    for r_ind, r in enumerate(records): \n",
    "\n",
    "        trans = r.seq[:-1] # ignores final stop codon\n",
    "        \n",
    "        if protein=='ORF1b': # ignore this masked region\n",
    "            if '*' in trans or 'X' in trans or '-' in trans: # doesnt seem to start with an M for anything\n",
    "                num_errs+=1\n",
    "                error_seq[r_ind] += 1\n",
    "                stop_errs.append(  (np.asarray( list(trans) )=='*').sum()  )\n",
    "        \n",
    "        elif '*' in trans or trans[0] != 'M' or 'X' in trans or '-' in trans: \n",
    "            num_errs+=1\n",
    "            error_seq[r_ind] += 1\n",
    "            stop_errs.append(  (np.asarray( list(trans) )=='*').sum()  )\n",
    "\n",
    "            #if p=='ORF14':\n",
    "            #print(trans)\n",
    "\n",
    "            if trans[0] != 'M':\n",
    "                m_errs += 1\n",
    "\n",
    "\n",
    "    print(protein, num_errs, num_errs/len(trans), m_errs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([10944.,     0., 11395.,     0.,     0.,   999.,     0.,   382.,\n",
       "            0.,    12.]),\n",
       " array([0. , 0.4, 0.8, 1.2, 1.6, 2. , 2.4, 2.8, 3.2, 3.6, 4. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD4CAYAAADsKpHdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAQaklEQVR4nO3dbYxcZ3nG8f+FHYIhTUnqTep63dqoFq1jiZJYrmkkhGqkuAThfCCSkSAWSmU1Ci20lZDNh0b9YCmVKkpTNaksQuMUSmoF1FgBt40MCFVKk25eqOMYN1tC423ceIES3LdQh7sf5ok0rGft2Z3dmY3z/0mjOXOf55m5/WTXl885M5NUFZIkvW7UDUiSlgYDQZIEGAiSpMZAkCQBBoIkqVk+6gbma+XKlbV27dpRtyFJryqPPfbYd6pqrNe+V20grF27lomJiVG3IUmvKkn+dbZ9njKSJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAa/iTyprbtbu/tLIXvvbt18/steW1D+PECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBr9HPIfiefEk6m0cIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAvoIhCSfSXIqyVNdtcuTPJTkmXZ/Wde+PUkmkxxPcl1X/ZokR9q+O5Kk1S9O8let/kiStQv8Z5Qk9aGfI4R7gG0zaruBw1W1HjjcHpNkA7ADuKrNuTPJsjbnLmAXsL7dXnnOm4H/qKqfB/4I+IP5/mEkSfN33kCoqq8D35tR3g7sb9v7gRu66vdV1UtV9SwwCWxOsgq4tKoerqoC7p0x55Xnuh/Y+srRgyRpeOZ7DeHKqjoJ0O6vaPXVwImucVOttrptz6z/2JyqOgO8CPxUrxdNsivJRJKJ6enpebYuSeploS8q9/qXfZ2jfq45Zxer9lXVpqraNDY2Ns8WJUm9zDcQXmingWj3p1p9CljTNW4ceL7Vx3vUf2xOkuXAT3L2KSpJ0iKbbyAcBHa27Z3AA131He2dQ+voXDx+tJ1WOp1kS7s+cNOMOa881/uBr7TrDJKkIVp+vgFJPg+8C1iZZAq4DbgdOJDkZuA54EaAqjqa5ADwNHAGuLWqXm5PdQuddyytAA61G8DdwF8kmaRzZLBjQf5kkqQ5OW8gVNUHZtm1dZbxe4G9PeoTwMYe9f+lBYokaXT8pLIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgYMhCS/neRokqeSfD7JG5JcnuShJM+0+8u6xu9JMpnkeJLruurXJDnS9t2RJIP0JUmau3kHQpLVwG8Bm6pqI7AM2AHsBg5X1XrgcHtMkg1t/1XANuDOJMva090F7ALWt9u2+fYlSZqfQU8ZLQdWJFkOvBF4HtgO7G/79wM3tO3twH1V9VJVPQtMApuTrAIuraqHq6qAe7vmSJKGZN6BUFX/Bvwh8BxwEnixqv4OuLKqTrYxJ4Er2pTVwImup5hqtdVte2b9LEl2JZlIMjE9PT3f1iVJPQxyyugyOv/qXwf8DPCmJB8815QetTpH/exi1b6q2lRVm8bGxubasiTpHAY5ZfRu4Nmqmq6q/wO+CPwK8EI7DUS7P9XGTwFruuaP0znFNNW2Z9YlSUM0SCA8B2xJ8sb2rqCtwDHgILCzjdkJPNC2DwI7klycZB2di8ePttNKp5Nsac9zU9ccSdKQLJ/vxKp6JMn9wOPAGeAJYB9wCXAgyc10QuPGNv5okgPA0238rVX1cnu6W4B7gBXAoXaTJA3RvAMBoKpuA26bUX6JztFCr/F7gb096hPAxkF6kSQNxk8qS5IAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQMFQpI3J7k/yTeTHEvyjiSXJ3koyTPt/rKu8XuSTCY5nuS6rvo1SY60fXckySB9SZLmbtAjhD8G/qaqfgF4G3AM2A0crqr1wOH2mCQbgB3AVcA24M4ky9rz3AXsAta327YB+5IkzdG8AyHJpcA7gbsBquqHVfV9YDuwvw3bD9zQtrcD91XVS1X1LDAJbE6yCri0qh6uqgLu7ZojSRqSQY4Q3gJMA3+e5Ikkn07yJuDKqjoJ0O6vaONXAye65k+12uq2PbN+liS7kkwkmZienh6gdUnSTIMEwnLgauCuqno78F+000Oz6HVdoM5RP7tYta+qNlXVprGxsbn2K0k6h0ECYQqYqqpH2uP76QTEC+00EO3+VNf4NV3zx4HnW328R12SNETzDoSq+nfgRJK3ttJW4GngILCz1XYCD7Ttg8COJBcnWUfn4vGj7bTS6SRb2ruLbuqaI0kakuUDzv9N4HNJXg98C/gwnZA5kORm4DngRoCqOprkAJ3QOAPcWlUvt+e5BbgHWAEcajdJ0hANFAhV9SSwqceurbOM3wvs7VGfADYO0oskaTB+UlmSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQsQCEmWJXkiyYPt8eVJHkryTLu/rGvsniSTSY4nua6rfk2SI23fHUkyaF+SpLlZiCOEjwLHuh7vBg5X1XrgcHtMkg3ADuAqYBtwZ5Jlbc5dwC5gfbttW4C+JElzMFAgJBkHrgc+3VXeDuxv2/uBG7rq91XVS1X1LDAJbE6yCri0qh6uqgLu7ZojSRqSQY8QPgV8HPhRV+3KqjoJ0O6vaPXVwImucVOttrptz6yfJcmuJBNJJqanpwdsXZLUbd6BkOS9wKmqeqzfKT1qdY762cWqfVW1qao2jY2N9fmykqR+LB9g7rXA+5K8B3gDcGmSzwIvJFlVVSfb6aBTbfwUsKZr/jjwfKuP96hLkoZo3kcIVbWnqsarai2di8VfqaoPAgeBnW3YTuCBtn0Q2JHk4iTr6Fw8frSdVjqdZEt7d9FNXXMkSUMyyBHCbG4HDiS5GXgOuBGgqo4mOQA8DZwBbq2ql9ucW4B7gBXAoXaTJA3RggRCVX0N+Frb/i6wdZZxe4G9PeoTwMaF6EWSND9+UlmSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJIaA0GSBBgIkqTGQJAkAQMEQpI1Sb6a5FiSo0k+2uqXJ3koyTPt/rKuOXuSTCY5nuS6rvo1SY60fXckyWB/LEnSXA1yhHAG+N2q+kVgC3Brkg3AbuBwVa0HDrfHtH07gKuAbcCdSZa157oL2AWsb7dtA/QlSZqHeQdCVZ2sqsfb9mngGLAa2A7sb8P2Aze07e3AfVX1UlU9C0wCm5OsAi6tqoerqoB7u+ZIkoZkQa4hJFkLvB14BLiyqk5CJzSAK9qw1cCJrmlTrba6bc+s93qdXUkmkkxMT08vROuSpGbgQEhyCfAF4GNV9YNzDe1Rq3PUzy5W7auqTVW1aWxsbO7NSpJmNVAgJLmIThh8rqq+2MovtNNAtPtTrT4FrOmaPg483+rjPeqSpCEa5F1GAe4GjlXVJ7t2HQR2tu2dwANd9R1JLk6yjs7F40fbaaXTSba057ypa44kaUiWDzD3WuBDwJEkT7baJ4DbgQNJbgaeA24EqKqjSQ4AT9N5h9KtVfVym3cLcA+wAjjUbpKkIZp3IFTV39P7/D/A1lnm7AX29qhPABvn24skaXB+UlmSBBgIkqTGQJAkAQaCJKkxECRJgIEgSWoMBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEnAYF9/LamHtbu/NLLX/vbt14/stfXq5xGCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNX10haWB+XceFwSMESRJgIEiSGgNBkgQsoUBIsi3J8SSTSXaPuh9Jeq1ZEoGQZBnwp8CvARuADyTZMNquJOm1ZUkEArAZmKyqb1XVD4H7gO0j7kmSXlNSVaPugSTvB7ZV1a+3xx8CfrmqPjJj3C5gV3v4VuD4PF9yJfCdec5dTPY1N/Y1d0u1N/uam0H6+rmqGuu1Y6l8DiE9amclVVXtA/YN/GLJRFVtGvR5Fpp9zY19zd1S7c2+5max+loqp4ymgDVdj8eB50fUiyS9Ji2VQPhHYH2SdUleD+wADo64J0l6TVkSp4yq6kySjwB/CywDPlNVRxfxJQc+7bRI7Gtu7Gvulmpv9jU3i9LXkrioLEkavaVyykiSNGIGgiQJuMAD4Xxfh5GOO9r+f0py9RLp611JXkzyZLv93pD6+kySU0memmX/qNbrfH0Nfb2SrEny1STHkhxN8tEeY4a+Xn32NYr1ekOSR5N8o/X1+z3GjGK9+ulrJL+P7bWXJXkiyYM99i38elXVBXmjc3H6X4C3AK8HvgFsmDHmPcAhOp+D2AI8skT6ehfw4AjW7J3A1cBTs+wf+nr12dfQ1wtYBVzdtn8C+Ocl8vPVT1+jWK8Al7Tti4BHgC1LYL366Wskv4/ttX8H+Mter78Y63UhHyH083UY24F7q+MfgDcnWbUE+hqJqvo68L1zDBnFevXT19BV1cmqerxtnwaOAatnDBv6evXZ19C1NfjP9vCidpv5jpZRrFc/fY1EknHgeuDTswxZ8PW6kANhNXCi6/EUZ/9i9DNmFH0BvKMdxh5KctUi99SvUaxXv0a2XknWAm+n86/LbiNdr3P0BSNYr3b640ngFPBQVS2J9eqjLxjNz9engI8DP5pl/4Kv14UcCP18HUZfX5mxwPp5zcfpfN/I24A/Af56kXvq1yjWqx8jW68klwBfAD5WVT+YubvHlKGs13n6Gsl6VdXLVfVLdL6JYHOSjTOGjGS9+uhr6OuV5L3Aqap67FzDetQGWq8LORD6+TqMUXxlxnlfs6p+8MphbFV9GbgoycpF7qsfS/IrRka1XkkuovOX7ueq6os9hoxkvc7X16h/vqrq+8DXgG0zdo3052u2vka0XtcC70vybTqnlX81yWdnjFnw9bqQA6Gfr8M4CNzUrtZvAV6sqpOj7ivJTydJ295M57/Tdxe5r36MYr3OaxTr1V7vbuBYVX1ylmFDX69++hrReo0leXPbXgG8G/jmjGGjWK/z9jWK9aqqPVU1XlVr6fwd8ZWq+uCMYQu+XkviqysWQ83ydRhJfqPt/zPgy3Su1E8C/w18eIn09X7gliRngP8BdlR7W8FiSvJ5Ou+oWJlkCriNzkW2ka1Xn32NYr2uBT4EHGnnnwE+AfxsV1+jWK9++hrFeq0C9qfzP8N6HXCgqh4c9e9jn32N5Pexl8VeL7+6QpIEXNinjCRJc2AgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJzf8D8HVPAfhlu1UAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(error_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12788"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(error_seq>0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23732"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(error_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_seq = (error_seq>0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing Belgian Sequences\n",
    "https://github.com/nextstrain/ncov/commit/6b7b822858ccc3e21ec279fc6afde30365d25aa0 They had 3 masked regions that need to be accounted for. May be more read quality issues too\n",
    "\n",
    "* This should no longer be a problem!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# getting the nt regions\\n\\nwith open(\\'nextstrain_covid19_ref_protein_pos.txt\\', \\'r\\') as f: \\n    lines = f.readlines()\\ntake_next_line = False\\nregions = []\\nproteins = []\\nfor l in lines:\\n    if \\'CDS \\' in l:\\n        regions.append(l.split(\\'CDS\\')[1].strip())\\n        take_next_line=True\\n        \\n    elif take_next_line:\\n        proteins.append(l.split(\\'gene=\\')[1].strip().strip(\\'\"\\'))\\n        take_next_line=False\\n        \\nprotein_regions = {p:r for p,r in zip(proteins, regions)}\\nprotein_regions'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# getting the nt regions\n",
    "\n",
    "with open('nextstrain_covid19_ref_protein_pos.txt', 'r') as f: \n",
    "    lines = f.readlines()\n",
    "take_next_line = False\n",
    "regions = []\n",
    "proteins = []\n",
    "for l in lines:\n",
    "    if 'CDS ' in l:\n",
    "        regions.append(l.split('CDS')[1].strip())\n",
    "        take_next_line=True\n",
    "        \n",
    "    elif take_next_line:\n",
    "        proteins.append(l.split('gene=')[1].strip().strip('\"'))\n",
    "        take_next_line=False\n",
    "        \n",
    "protein_regions = {p:r for p,r in zip(proteins, regions)}\n",
    "protein_regions'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'belgian_masked_sites = [ \\n               13402-266,\\n               24389-21563, \\n               24390-21563]\\nbelgian_masked_sites = np.asarray(belgian_masked_sites)/3\\nbelgian_masked_sites'"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''belgian_masked_sites = [ \n",
    "               13402-266,\n",
    "               24389-21563, \n",
    "               24390-21563]\n",
    "belgian_masked_sites = np.asarray(belgian_masked_sites)/3\n",
    "belgian_masked_sites'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"the_proteins = ['ORF1a', 'S', 'S']\\n#for prot, pos in zip(the_proteins,belgian_masked_sites )\\nmasked_dict = {'ORF1a':4378, 'S':942 }\\nmasked_dict\""
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''the_proteins = ['ORF1a', 'S', 'S']\n",
    "#for prot, pos in zip(the_proteins,belgian_masked_sites )\n",
    "masked_dict = {'ORF1a':4378, 'S':942 }\n",
    "masked_dict'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'for aa_file in amino_acid_files:\\n    protein = aa_file.split(\\'_\\')[-1].split(\\'.\\')[0]\\n    print(\\'protein:\\', protein)\\n    old_aa_file = \\'../PUFFIN/proteins/v4/aligned_protein_\\'+protein+\\'.fasta\\'\\n\\n    with open(old_aa_file, \"rt\") as handle:\\n        old_records = list(SeqIO.parse(handle, \"fasta\"))\\n        \\n    n_belg = 0\\n    no_nodes = []\\n    for r in old_records: \\n        if \\'Belgium\\' in r.id:\\n            n_belg+=1\\n        if \\'NODE_\\' not in r.id:\\n            no_nodes.append(r)\\n    old_records = no_nodes\\n    print(\\'reference sequences number belgian\\', n_belg)\\n        \\n    for ind, r in enumerate(old_records): \\n        if r.id == \\'Wuhan/IPBCAMS-WH-01/2019\\':#\\'Wuhan-Hu-1/2019\\':#\\'Wuhan/WH01/2019\\':#\\'Wuhan/IPBCAMS-WH-01/2019\\':\\n            print(ind)\\n            ref_seq_ind = ind\\n    ref_seq_old = str(old_records[ref_seq_ind].seq)[:-1]\\n    # this sequence does not have these same masking positions applied. \\n    \\n    curr_aa_file = \\'../PUFFIN/proteins/v6/cleaned_aligned_protein_\\'+protein+\\'.fasta\\'\\n    with open(curr_aa_file, \"rt\") as handle:\\n        records = list(SeqIO.parse(handle, \"fasta\"))\\n\\n    belg_seqs=[]\\n    for ind, r in enumerate(records): \\n        if \\'Belgium\\' in r.id:\\n            belg_seqs.append(r.seq)\\n            \\n    print(\\'num belg seqs\\', len(belg_seqs))\\n    \\n    if protein in masked_dict.keys():\\n        mask_pos = masked_dict[protein]\\n        print(\\'reference value\\',ref_seq_old[mask_pos] )\\n        \\n        old_records= np.asarray(old_records)\\n        unique, counts = np.unique(old_records[:, mask_pos], return_counts=True)\\n        count_dict = dict(zip(unique, counts))\\n        print(\\'reference values\\', count_dict)\\n        \\n        belg_seqs = np.asarray(belg_seqs)\\n        unique, counts = np.unique(belg_seqs[:, mask_pos], return_counts=True)\\n        count_dict = dict(zip(unique, counts))\\n        print(\\'belgian masked values\\', count_dict)\\n        \\n        \\n        records = np.asarray(records)\\n        unique, counts = np.unique(records[:, mask_pos], return_counts=True)\\n        count_dict = dict(zip(unique, counts))\\n        print(\\'all masked sequences\\', count_dict)'"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''for aa_file in amino_acid_files:\n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    print('protein:', protein)\n",
    "    old_aa_file = '../PUFFIN/proteins/v4/aligned_protein_'+protein+'.fasta'\n",
    "\n",
    "    with open(old_aa_file, \"rt\") as handle:\n",
    "        old_records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "        \n",
    "    n_belg = 0\n",
    "    no_nodes = []\n",
    "    for r in old_records: \n",
    "        if 'Belgium' in r.id:\n",
    "            n_belg+=1\n",
    "        if 'NODE_' not in r.id:\n",
    "            no_nodes.append(r)\n",
    "    old_records = no_nodes\n",
    "    print('reference sequences number belgian', n_belg)\n",
    "        \n",
    "    for ind, r in enumerate(old_records): \n",
    "        if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "            print(ind)\n",
    "            ref_seq_ind = ind\n",
    "    ref_seq_old = str(old_records[ref_seq_ind].seq)[:-1]\n",
    "    # this sequence does not have these same masking positions applied. \n",
    "    \n",
    "    curr_aa_file = '../PUFFIN/proteins/v6/cleaned_aligned_protein_'+protein+'.fasta'\n",
    "    with open(curr_aa_file, \"rt\") as handle:\n",
    "        records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "    belg_seqs=[]\n",
    "    for ind, r in enumerate(records): \n",
    "        if 'Belgium' in r.id:\n",
    "            belg_seqs.append(r.seq)\n",
    "            \n",
    "    print('num belg seqs', len(belg_seqs))\n",
    "    \n",
    "    if protein in masked_dict.keys():\n",
    "        mask_pos = masked_dict[protein]\n",
    "        print('reference value',ref_seq_old[mask_pos] )\n",
    "        \n",
    "        old_records= np.asarray(old_records)\n",
    "        unique, counts = np.unique(old_records[:, mask_pos], return_counts=True)\n",
    "        count_dict = dict(zip(unique, counts))\n",
    "        print('reference values', count_dict)\n",
    "        \n",
    "        belg_seqs = np.asarray(belg_seqs)\n",
    "        unique, counts = np.unique(belg_seqs[:, mask_pos], return_counts=True)\n",
    "        count_dict = dict(zip(unique, counts))\n",
    "        print('belgian masked values', count_dict)\n",
    "        \n",
    "        \n",
    "        records = np.asarray(records)\n",
    "        unique, counts = np.unique(records[:, mask_pos], return_counts=True)\n",
    "        count_dict = dict(zip(unique, counts))\n",
    "        print('all masked sequences', count_dict)'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'unique, counts = np.unique(belg_seqs[:, mask_pos], return_counts=True)\\ncount_dict = dict(zip(unique, counts))\\ncount_dict'"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''unique, counts = np.unique(belg_seqs[:, mask_pos], return_counts=True)\n",
    "count_dict = dict(zip(unique, counts))\n",
    "count_dict'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving out these cleaned sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # write out the dropped sequences. \n",
    "# from Bio import SeqIO\n",
    "# for aa_file in amino_acid_files: \n",
    "    \n",
    "#     protein = aa_file.split('_')[-1].split('.')[0]\n",
    "#     #print('protein:', protein)\n",
    "#     with open(aa_file, \"rt\") as handle:\n",
    "#         records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "        \n",
    "#     # getting rid of the node sequences!\n",
    "#     no_nodes = []\n",
    "#     for r in records: \n",
    "#         if 'NODE_' not in r.id:\n",
    "#             no_nodes.append(r)\n",
    "#     records = no_nodes\n",
    "    \n",
    "#     clean_records = []\n",
    "#     for r_ind, r in enumerate(records): \n",
    "#         if drop_seq[r_ind]==False:\n",
    "#             clean_records.append(r)\n",
    "#     print(len(clean_records))\n",
    "\n",
    "#     with open(\"../PUFFIN/proteins/v6_cleaned/aligned_protein_\"+protein+\".fasta\", \"w\") as output_handle:\n",
    "#         SeqIO.write(clean_records, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# amino_acid_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the Entropy in a Parallelized Fashion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def computeWindows(aa_file):\n",
    "    print('computeWindows!!!')\n",
    "    print('aa_file = ', aa_file) ## Alex\n",
    "    \n",
    "    def entropy_calc(x):\n",
    "        # where x is a list of numbers. \n",
    "        summ = 0\n",
    "        nc = np.sum(x)\n",
    "        for e in x: \n",
    "            p = e/nc\n",
    "            summ += p*np.log2(p)\n",
    "        return -summ\n",
    "    \n",
    "    window_sizes = list(range(8,12)) + list(range(13,26)) # why were sliding windows of 12 excluded? -- Alex\n",
    "    #df = dict() # will store the results. \n",
    "    protein_res = []\n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    print('protein:', protein)\n",
    "    with open(aa_file, \"rt\") as handle:\n",
    "        records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "    \n",
    "    \n",
    "    # getting rid of the node sequences!\n",
    "    no_nodes = []\n",
    "    for r in records: \n",
    "        if 'NODE_' not in r.id:\n",
    "            no_nodes.append(r)\n",
    "    records = no_nodes\n",
    "    print('size of records', len(records))\n",
    "    # getting the reference sequence\n",
    "    for ind, r in enumerate(records): \n",
    "        if r.id == 'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "            ref_seq_ind = ind\n",
    "            \n",
    "    ref_seq = str(records[ref_seq_ind].seq)\n",
    "    \n",
    "    seqs = np.array(records)\n",
    "    \n",
    "    if seqs[ref_seq_ind, -1]=='*': # ignore the stop code at the end. \n",
    "        print('removing the stop codon at the end', protein)\n",
    "        ref_seq = ref_seq[:-1]\n",
    "        seqs = seqs[:, :-1]\n",
    "    \n",
    "    for window in window_sizes:\n",
    "        print('window size', window)\n",
    "        \n",
    "        window_res = []\n",
    "        \n",
    "        # get epitope based column slices. gives a list of epi columns. \n",
    "        epi_columns = [seqs[:,i:i+window] for i in range(seqs.shape[1]-window+1)]\n",
    "    \n",
    "        for col_ind, col in enumerate(epi_columns):\n",
    "            # count all of the unique epitopes here. \n",
    "            \n",
    "            #first need to convert each of the columns into strings:\n",
    "            col = np.asarray(  [  ''.join(col[i,:]) for i in range(col.shape[0]) ])\n",
    "            unique, counts = np.unique(col, return_counts=True) \n",
    "            ent = entropy_calc(counts)\n",
    "            \n",
    "            # useful for percentages\n",
    "            ref_epitope = col[ref_seq_ind]\n",
    "            count_dict = dict(zip(unique, counts))\n",
    "            # -1 for no self count. percentage mutated. \n",
    "            perc = 1 - ((count_dict[ref_epitope]-1)/(np.sum(counts)-1))\n",
    "            \n",
    "            # start pos, window size, entropy\n",
    "            window_res.append([protein, ref_epitope, col_ind, window, ent, perc])\n",
    "            \n",
    "        protein_res += window_res\n",
    "        \n",
    "    df = pd.DataFrame(protein_res)\n",
    "    df.columns = ['protein', 'epitope', 'start_pos', 'epi_len', 'entropy', 'perc_mutated']\n",
    "#     df.to_csv('../data/processed/epitope_calcs/Epitope_calc_'+protein+'.csv' , index=False)\n",
    "    df.to_csv('data/processed/epitope_calcs_larger/Epitope_calc_'+protein+'.csv' , index=False)\n",
    "\n",
    "        \n",
    "    #df[protein] = protein_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/larger_version_modified/aligned_protein_E.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_M.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_N.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF10.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF1a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF1b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF3a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF6.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF7a.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF7b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF8.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_ORF9b.fasta',\n",
       " 'data/larger_version_modified/aligned_protein_S.fasta']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Check that these are the larger_version_modified files\n",
    "amino_acid_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # check these are the right AA files: \n",
    "# # amino_acid_files = !ls ../PUFFIN/proteins/v6_cleaned/*_protein_*\n",
    "# amino_acid_files = !ls data/v6_cleaned/*_protein_*\n",
    "\n",
    "# amino_acid_files\n",
    "# # len(amino_acid_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting pooling!!!\n",
      "computeWindows!!!computeWindows!!!computeWindows!!!\n",
      "\n",
      "computeWindows!!!\n",
      "aa_file = computeWindows!!!aa_file = aa_file = computeWindows!!!\n",
      "computeWindows!!!  \n",
      " aa_file = data/larger_version_modified/aligned_protein_M.fasta\n",
      "\n",
      "aa_file = data/larger_version_modified/aligned_protein_E.fasta\n",
      " aa_file = data/larger_version_modified/aligned_protein_N.fasta\n",
      "aa_file =  protein:data/larger_version_modified/aligned_protein_ORF10.fasta \n",
      "data/larger_version_modified/aligned_protein_ORF1a.fastaprotein:  \n",
      "data/larger_version_modified/aligned_protein_ORF1b.fastaprotein:\n",
      " data/larger_version_modified/aligned_protein_ORF3a.fastaMprotein:E\n",
      " protein:\n",
      "\n",
      " \n",
      " protein:Nprotein:ORF1aORF10 \n",
      " \n",
      "\n",
      "ORF1bORF3a\n",
      "\n",
      "size of records size of records12789 \n",
      "12789\n",
      "size of records 12789\n",
      "size of records 12789\n",
      "size of records 12789\n",
      "size of records removing the stop codon at the end12789\n",
      " ORF10\n",
      "window size 8\n",
      "removing the stop codon at the end E\n",
      "size of recordswindow size  127898\n",
      "\n",
      "removing the stop codon at the end M\n",
      "window size 8\n",
      "window size 9\n",
      "removing the stop codon at the end ORF3a\n",
      "window size 8\n",
      "window size 10\n",
      "removing the stop codon at the end N\n",
      "window size 8\n",
      "window size 9\n",
      "window size 11\n",
      "window size 13\n",
      "window size 10\n",
      "window size 14\n",
      "window size 15\n",
      "window size 11\n",
      "window size 9\n",
      "window size 16\n",
      "window size 17\n",
      "window size 9\n",
      "window size 13\n",
      "window size 18\n",
      "window size 19\n",
      "window size 14\n",
      "window size 20\n",
      "window size 21\n",
      "window size 22\n",
      "removing the stop codon at the end ORF1b\n",
      "window size 8\n",
      "window size 9\n",
      "window size 10\n",
      "window size 15\n",
      "window size 23\n",
      "window size 24\n",
      "window size 25\n",
      "window size 16\n",
      "window size 10\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_ORF6.fasta\n",
      "protein: ORF6\n",
      "size of records 12789\n",
      "removing the stop codon at the end ORF6\n",
      "window size 8\n",
      "window size 17\n",
      "window size 9\n",
      "window size 11\n",
      "window size 10\n",
      "window size 8\n",
      "window size 18\n",
      "window size 11\n",
      "window size 13\n",
      "window size 19\n",
      "window size 11\n",
      "window size 10\n",
      "window size 14\n",
      "window size 13\n",
      "window size 20\n",
      "window size 15\n",
      "window size 16\n",
      "window size 21\n",
      "window size 17\n",
      "window size 22\n",
      "window size 18\n",
      "window size 13\n",
      "window size 14\n",
      "window size 19\n",
      "window size 23\n",
      "window size 20\n",
      "window size 11\n",
      "window size 24\n",
      "window size 21\n",
      "window size 22\n",
      "window size 25\n",
      "window size 15\n",
      "window size 14\n",
      "window size 23\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_ORF7a.fasta\n",
      "protein: ORF7a\n",
      "size of records 12789\n",
      "removing the stop codon at the end ORF7a\n",
      "window size 8\n",
      "window size 24\n",
      "window size 25\n",
      "window size 9\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_ORF7b.fasta\n",
      "protein: ORF7b\n",
      "size of records 12789\n",
      "removing the stop codon at the end ORF7b\n",
      "window size 8\n",
      "window size 13\n",
      "window size 16\n",
      "window size 10\n",
      "window size 9\n",
      "window size 10\n",
      "window size 15\n",
      "window size 11\n",
      "window size 13\n",
      "window size 11\n",
      "window size 14\n",
      "window size 15\n",
      "window size 16\n",
      "window size 13\n",
      "window size 17\n",
      "window size 17\n",
      "window size 18\n",
      "window size 19\n",
      "window size 14\n",
      "window size 20\n",
      "window size 16\n",
      "window size 21\n",
      "window size 22\n",
      "window size 14\n",
      "window size 23\n",
      "window size 15\n",
      "window size 24\n",
      "window size 18\n",
      "window size 25\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_ORF8.fasta\n",
      "protein: ORF8\n",
      "size of records 12789\n",
      "removing the stop codon at the end ORF8\n",
      "window size 8\n",
      "window size 16\n",
      "window size 9\n",
      "window size 17\n",
      "window size 17\n",
      "window size 10\n",
      "window size 19\n",
      "window size 18\n",
      "window size 11\n",
      "window size 15\n",
      "window size 9\n",
      "window size 13\n",
      "window size 19\n",
      "window size 18\n",
      "window size 14\n",
      "window size 20\n",
      "window size 20\n",
      "window size 15\n",
      "window size 21\n",
      "window size 16\n",
      "window size 16\n",
      "window size 19\n",
      "window size 21\n",
      "window size 22\n",
      "window size 17\n",
      "window size 18\n",
      "window size 23\n",
      "window size 22\n",
      "window size 19\n",
      "window size 20\n",
      "window size 24\n",
      "window size 17\n",
      "window size 20\n",
      "window size 25\n",
      "window size 23\n",
      "window size 21\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_ORF9b.fasta\n",
      "protein: ORF9b\n",
      "size of records 12789\n",
      "window size 21\n",
      "removing the stop codon at the end ORF9b\n",
      "window size 8\n",
      "window size 22\n",
      "window size 9\n",
      "window size 10\n",
      "window size 11\n",
      "window size 23\n",
      "window size 24\n",
      "window size 18\n",
      "window size 13\n",
      "window size 9\n",
      "window size 22\n",
      "window size 24\n",
      "window size 14\n",
      "window size 15\n",
      "window size 25\n",
      "window size 25\n",
      "window size 16\n",
      "window size 17\n",
      "computeWindows!!!\n",
      "aa_file =  data/larger_version_modified/aligned_protein_S.fasta\n",
      "protein: S\n",
      "size of records 12789\n",
      "window size 23\n",
      "window size 18\n",
      "window size 19\n",
      "removing the stop codon at the end S\n",
      "window size window size8 \n",
      "19\n",
      "window size 20\n",
      "window size 10\n",
      "window size 21\n",
      "window size 24\n",
      "window size 22\n",
      "window size 23\n",
      "window size 20\n",
      "window size 24\n",
      "window size 25\n",
      "window size 25\n",
      "window size 9\n",
      "window size 21\n",
      "window size 22\n",
      "window size 10\n",
      "window size 23\n",
      "window size 11\n",
      "window size 24\n",
      "window size 10\n",
      "window size 11\n",
      "window size 25\n",
      "window size 13\n",
      "window size 13\n",
      "window size 14\n",
      "window size 11\n",
      "window size 15\n",
      "window size 14\n",
      "window size 16\n",
      "window size 17\n",
      "window size 15\n",
      "window size 13\n",
      "window size 18\n",
      "window size 19\n",
      "window size 16\n",
      "window size 20\n",
      "window size 14\n",
      "window size 21\n",
      "window size 17\n",
      "window size 22\n",
      "window size 15\n",
      "window size 18\n",
      "window size 23\n",
      "window size 24\n",
      "window size 19\n",
      "window size 25\n",
      "window size 16\n",
      "window size 20\n",
      "window size 17\n",
      "window size 21\n",
      "window size 22\n",
      "window size 18\n",
      "window size 23\n",
      "window size 19\n",
      "window size 24\n",
      "window size 20\n",
      "window size 25\n",
      "window size 21\n",
      "window size 22\n",
      "window size 23\n",
      "window size 24\n",
      "window size 25\n",
      "all processes are done!, trying to join together all of the files\n",
      "Total run time in minutes: 103.48027988274892\n"
     ]
    }
   ],
   "source": [
    "# from multiprocessing import Process, Queue, cpu_count, Pool\n",
    "from multiprocess import Process, Queue, cpu_count, Pool\n",
    "import time \n",
    "ncores = 7\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "### uncomment this to run it, commented for safety reasons -- Alex\n",
    "# multicore generate new samples\n",
    "print('Starting pooling!!!')\n",
    "p = Pool(ncores)\n",
    "p.map(computeWindows, amino_acid_files)\n",
    "p.close()\n",
    "\n",
    "print('all processes are done!, trying to join together all of the files')\n",
    "\n",
    "for ind, aa_file in enumerate(amino_acid_files):\n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    \n",
    "#     temp = pd.read_csv('../data/processed/epitope_calcs/Epitope_calc_'+protein+'.csv')\n",
    "    temp = pd.read_csv('data/processed/epitope_calcs_larger/Epitope_calc_'+protein+'.csv')\n",
    "    \n",
    "    if ind == 0:\n",
    "        df = temp\n",
    "    else: \n",
    "        df = df.append(temp)\n",
    "    \n",
    "print('Total run time in minutes: '+str((time.time()-start_time)/60))\n",
    "\n",
    "# df.to_csv('../data/processed/parallelized_epitope_entropies.csv', index=False)\n",
    "df.to_csv('data/processed/parallelized_epitope_entropies_larger.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Can ignore all of the following now. \n",
    "\n",
    "### Looking at the mutation differences between Hu-1 and original sequence\n",
    "\n",
    "* This analysis motivated my suggestion that we should in fact be using the Hu-1 sequence as we were missing a couple potential epitopes we could have chosen. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.000213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0       E  MYSFVSEE          0        8  0.002908      0.000213\n",
       "1       E  YSFVSEET          1        8  0.008297      0.000640\n",
       "2       E  SFVSEETG          2        8  0.008297      0.000640\n",
       "3       E  FVSEETGT          3        8  0.008297      0.000640\n",
       "4       E  VSEETGTL          4        8  0.008297      0.000640"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#df = pd.read_csv('../data/processed/epitope_entropies.csv')\n",
    "#df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164786, 6)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.000213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0       E  MYSFVSEE          0        8  0.002908      0.000213\n",
       "1       E  YSFVSEET          1        8  0.008297      0.000640\n",
       "2       E  SFVSEETG          2        8  0.008297      0.000640\n",
       "3       E  FVSEETGT          3        8  0.008297      0.000640\n",
       "4       E  VSEETGTL          4        8  0.008297      0.000640"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAQfklEQVR4nO3dfYxcV33G8e+D84bESxK8VNR2sKGmwkJRQ1dpUKo2BQpOgmz+SFtHIAKNYqlt2kYgKiOqAKkqBZBKqZqWuoB4EeCEtAILDBaCpEiIhGxISElSw2JCszhqTEhCKyAv9Nc/5jod7c7uzq7HHu/Z70da7b3nnrnzu8e7j++euTM3VYUkaeV72rgLkCSNhoEuSY0w0CWpEQa6JDXCQJekRpw0rideu3Ztbdy4cVxPL0kr0u233/6jqpoYtG1sgb5x40ampqbG9fSStCIl+cF825xykaRGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDVi0TcWJfkw8Brgwap6yYDtAd4PXAT8FHhjVX1z1IVKq8XGXZ+fd9t911488sepHcOcoX8E2LrA9guBzd3XTuAfj74saXVaKJQX2r7cx6ktiwZ6VX0V+PECXbYDH6ueW4DTkzxvVAVKkoYzijn0dcD9feszXdscSXYmmUoydfjw4RE8tSTpiFEEega0DbxRaVXtrqrJqpqcmBj4YWGSpGUaRaDPABv61tcDh0awX0nSEowi0PcCb0jPecCjVfXACPYrrTqLXY0y3/blPk5tGeayxU8BFwBrk8wA7wBOBqiqDwD76F2yOE3vssU3HatipdVgueFraGvRQK+qSxfZXsCfjKwiSdKy+E5RSWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGDBXoSbYmOZBkOsmuAdvPSnJTkjuS3JXkotGXKklayKKBnmQNcB1wIbAFuDTJllnd/hK4oarOAXYA/zDqQiVJCxvmDP1cYLqqDlbV48AeYPusPgU8q1t+NnBodCVKkoYxTKCvA+7vW5/p2vq9E3h9khlgH/Cng3aUZGeSqSRThw8fXka5kqT5DBPoGdBWs9YvBT5SVeuBi4CPJ5mz76raXVWTVTU5MTGx9GolSfMaJtBngA196+uZO6VyOXADQFV9HTgNWDuKAiVJwxkm0G8DNifZlOQUei967p3V5z+BVwAkeTG9QHdORZKOo0UDvaqeBK4E9gP30rua5e4k1yTZ1nV7C3BFkm8BnwLeWFWzp2UkScfQScN0qqp99F7s7G+7um/5HuD80ZYmSVoK3ykqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGjFUoCfZmuRAkukku+bp8/tJ7klyd5JPjrZMSdJiTlqsQ5I1wHXA7wIzwG1J9lbVPX19NgNvA86vqoeTPPdYFSxJGmyYM/RzgemqOlhVjwN7gO2z+lwBXFdVDwNU1YOjLVOStJhhAn0dcH/f+kzX1u9FwIuSfC3JLUm2DtpRkp1JppJMHT58eHkVS5IGGibQM6CtZq2fBGwGLgAuBT6Y5PQ5D6raXVWTVTU5MTGx1FolSQsYJtBngA196+uBQwP6fLaqnqiq7wMH6AW8JOk4GSbQbwM2J9mU5BRgB7B3Vp/PAL8DkGQtvSmYg6MsVJK0sEUDvaqeBK4E9gP3AjdU1d1Jrkmyreu2H3goyT3ATcBbq+qhY1W0JGmuVM2eDj8+Jicna2pqaizPLUkrVZLbq2py0DbfKSpJjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1IihAj3J1iQHkkwn2bVAv0uSVJLJ0ZUoSRrGooGeZA1wHXAhsAW4NMmWAf2eCfwZcOuoi5QkLW6YM/RzgemqOlhVjwN7gO0D+v0V8B7g5yOsT5I0pGECfR1wf9/6TNf2lCTnABuq6nML7SjJziRTSaYOHz685GIlSfMbJtAzoK2e2pg8DXgf8JbFdlRVu6tqsqomJyYmhq9SkrSoYQJ9BtjQt74eONS3/kzgJcDNSe4DzgP2+sKoJB1fwwT6bcDmJJuSnALsAPYe2VhVj1bV2qraWFUbgVuAbVU1dUwqliQNtGigV9WTwJXAfuBe4IaqujvJNUm2HesCJUnDOWmYTlW1D9g3q+3qefpecPRlSZKWyneKSlIjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJasRJ4y7geNu46/Pzbrvv2ouPYyWSNFqr6gx9oTAfZrsknchWVaBLUssMdElqxFCBnmRrkgNJppPsGrD9zUnuSXJXki8nef7oS5UkLWTRQE+yBrgOuBDYAlyaZMusbncAk1V1NnAj8J5RFypJWtgwZ+jnAtNVdbCqHgf2ANv7O1TVTVX10271FmD9aMscjcWuYvEqF0kr2TCXLa4D7u9bnwF+Y4H+lwNfGLQhyU5gJ8BZZ501ZImjZWhLatUwZ+gZ0FYDOyavByaB9w7aXlW7q2qyqiYnJiaGr1KStKhhztBngA196+uBQ7M7JXkl8Hbgt6vqsdGUJ0ka1jBn6LcBm5NsSnIKsAPY298hyTnAPwHbqurB0ZcpSVrMooFeVU8CVwL7gXuBG6rq7iTXJNnWdXsv8Azg00nuTLJ3nt1Jko6RoT7Lpar2AftmtV3dt/zKEdclSVoi3ykqSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGrHqbkEn6cQ1313D/Aym4XiGLumEsNAtIL095HAMdElj97p//vq4S2iCgS5p7L72vR+Pu4QmGOiS1AgDXZIaYaBLGrvzX3jmuEtogoEuaew+ccXLFgx1L1scjtehSzohfOKKl427hBXPM3RJaoSBLkmNMNAlqREGuiQ1wkCXpEZ4lcsAgz4IyMumJJ3oPEOfZb5PdfPT3iSd6Ax0SWqEUy5qzkJ/TTl1ppZ5ht7nM3f8cNwl6CgtNjXm1Jla1vwZ+qZdn6f61gN8f56ztKuuv/O41HS0PnPHD3nv/gMceuRn/PLpT+etr/5VXnvOunGXJWnMVnSgL3b/wdlhDlBd+3yhvtjzjeJP9tf989cX/ED/+669eOgzyR8+8jOuuv5Orrr+TqcTpD6r8f6kKzbQF7v/4H3XXjwnzI8oRhfOS7VYmMPypwWWc0zzPVf/XzKzaz7/hWf6QUpjMOrXBpb77zrsz+c4g3OYfGhRquaLvWNrcnKypqamlv34cc+FLnYWPd8PzPGu+0gdx+J5h/mlOPsdX+Qnj/1i2Y9fqqUe51LHZ1Q1L/Reh+X+Wy2ltmF/do/25+a0NeE//vqio9rHMLXMPvbF6l7q7+eJ9B9AkturanLgtmECPclW4P3AGuCDVXXtrO2nAh8Dfh14CPiDqrpvoX0uJ9DHHeKSNErPOnUNd71r65Ies1CgL3qVS5I1wHXAhcAW4NIkW2Z1uxx4uKp+BXgf8O4lVTgEw1xSa37y2C84+x1fHNn+hplDPxeYrqqDAEn2ANuBe/r6bAfe2S3fCPx9ktRCp/8HDsAFFwxd6J6DDw3dV5JWlH+7dvE+QxjmOvR1wP196zNd28A+VfUk8CjwnNk7SrIzyVSSqSeeeGJ5FUuSBhrmDD0D2mafeQ/Th6raDeyG3hw6N988xNP37HDKRVKjlvSiawbFbc8wZ+gzwIa+9fXAofn6JDkJeDaw8LV5kiSedeqake1rmEC/DdicZFOSU4AdwN5ZffYCl3XLlwBfWXD+fBlOpMuGJGkUlnOVy0IWnXKpqieTXAnsp3fZ4oer6u4k1wBTVbUX+BDw8STT9M7Md4yswj6GuiTNb6h3ilbVPmDfrLar+5Z/DvzeaEuTJC2Fn7YoSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjxnaDiySHgR8s8+FrgR+NsJxWOC5zOSZzOSZzraQxeX5VTQzaMLZAPxpJpub7gPfVzHGZyzGZyzGZq5UxccpFkhphoEtSI1ZqoO8edwEnKMdlLsdkLsdkribGZEXOoUuS5lqpZ+iSpFkMdElqxIoL9CRbkxxIMp1k17jrOZaSfDjJg0m+3dd2ZpIvJflu9/2Mrj1J/q4bl7uSvLTvMZd1/b+b5LJBz7VSJNmQ5KYk9ya5O8mfd+2rdlySnJbkG0m+1Y3Ju7r2TUlu7Y7v+u6OYyQ5tVuf7rZv7NvX27r2A0lePZ4jGp0ka5LckeRz3XrbY1JVK+aL3h2Tvge8ADgF+BawZdx1HcPj/S3gpcC3+9reA+zqlncB7+6WLwK+QO+G3ecBt3btZwIHu+9ndMtnjPvYjmJMnge8tFt+JvAdYMtqHpfu2J7RLZ8M3Nod6w3Ajq79A8Afdct/DHygW94BXN8tb+l+p04FNnW/a2vGfXxHOTZvBj4JfK5bb3pMVtoZ+rnAdFUdrKrHgT3A9jHXdMxU1VeZe7Pt7cBHu+WPAq/ta/9Y9dwCnJ7kecCrgS9V1Y+r6mHgS8DobmJ4nFXVA1X1zW75v4F7gXWs4nHpju1/utWTu68CXg7c2LXPHpMjY3Uj8Iok6dr3VNVjVfV9YJre79yKlGQ9cDHwwW49ND4mKy3Q1wH3963PdG2ryS9V1QPQCzfguV37fGPT7Jh1fxafQ++MdFWPSze1cCfwIL3/nL4HPFJVT3Zd+o/vqWPvtj8KPIfGxgT4W+AvgP/t1p9D42Oy0gI9A9q87rJnvrFpcsySPAP4F+CqqvrJQl0HtDU3LlX1i6r6NWA9vTPIFw/q1n1vfkySvAZ4sKpu728e0LWpMVlpgT4DbOhbXw8cGlMt4/Jf3ZQB3fcHu/b5xqa5MUtyMr0w/0RV/WvXvOrHBaCqHgFupjeHfnqSIzeC7z++p4692/5selN7LY3J+cC2JPfRm5p9Ob0z9qbHZKUF+m3A5u6V6lPovXixd8w1HW97gSNXZFwGfLav/Q3dVR3nAY92Uw/7gVclOaO78uNVXduK1M1rfgi4t6r+pm/Tqh2XJBNJTu+Wnw68kt5rCzcBl3TdZo/JkbG6BPhK9V4B3Avs6K742ARsBr5xfI5itKrqbVW1vqo20suJr1TV62h9TMb9quxSv+hdtfAdenOEbx93Pcf4WD8FPAA8Qe9M4XJ683pfBr7bfT+z6xvgum5c/h2Y7NvPH9J7MWcaeNO4j+sox+Q36f3JexdwZ/d10WoeF+Bs4I5uTL4NXN21v4Be+EwDnwZO7dpP69anu+0v6NvX27uxOgBcOO5jG9H4XMD/X+XS9Jj41n9JasRKm3KRJM3DQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmN+D82cLyUAmmwjQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(df[df.protein=='ORF1a'].start_pos, df[df.protein=='ORF1a'].perc_mutated)\n",
    "plt.axhline(0.001, c='red')\n",
    "#plt.xlim(2690, 2727)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mutation positions = mut_pos = [2707, 2907]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_pos = [2707, 2907]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(74549, 6)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf1a = df[df.protein=='ORF1a']\n",
    "orf1a.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/trentonbricken/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "orf1a['crosses_mutation'] = 0\n",
    "for pos in mut_pos:\n",
    "\n",
    "    seq_start = orf1a['start_pos']\n",
    "    seq_end = orf1a['start_pos']+(orf1a['epi_len']-1)\n",
    "    \n",
    "    in_region = np.logical_and(pos >= seq_start,pos <= seq_end)\n",
    "    \n",
    "    orf1a.loc[in_region,'crosses_mutation'] = 1\n",
    "    #apply protein mask and then epitope mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(570, 7)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf1a[orf1a.crosses_mutation.astype(bool)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "570"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(orf1a[orf1a.crosses_mutation.astype(bool)].perc_mutated > 0.001).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "below_thresh_mask = orf1a[orf1a.crosses_mutation.astype(bool)].perc_mutated <= 0.001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "below_thresh_mask.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>crosses_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [protein, epitope, start_pos, epi_len, entropy, perc_mutated, crosses_mutation]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf1a[orf1a.crosses_mutation.astype(bool)][below_thresh_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=int64)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf1a[orf1a.crosses_mutation.astype(bool)][below_thresh_mask].start_pos.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# epitopes that are mutated according to our data but not the other papers. \n",
    "danger_epitopes = orf1a[orf1a.crosses_mutation.astype(bool)][below_thresh_mask].epitope.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pickle \n",
    "other_epitopes = pickle.load(open('../data/processed/Other_Methods_Proposed_Peptides.pkl', 'rb'))\n",
    "len(other_epitopes.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing MHC_1_Grifoni-LaJolla ['KLIEYTDFA']\n",
      "Missing MHC_1_Nerli-UCSC ['SKLIEYTDFA', 'KLIEYTDFAT']\n"
     ]
    }
   ],
   "source": [
    "for k, v in other_epitopes.items():\n",
    "    if len(list(set(danger_epitopes).intersection(set(v))))>0:\n",
    "        print('Missing', k, list(set(danger_epitopes).intersection(set(v))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is only the second mutation position that is below the mutation threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"for ind, aa_file in enumerate(amino_acid_files):\\n    protein = aa_file.split('_')[-1].split('.')[0]\\n    \\n    temp = pd.read_csv('../data/processed/epitope_calcs/Epitope_calc_'+protein+'.csv')\\n    \\n    if ind == 0:\\n        df = temp\\n    else: \\n        df = df.append(temp)\""
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''for ind, aa_file in enumerate(amino_acid_files):\n",
    "    protein = aa_file.split('_')[-1].split('.')[0]\n",
    "    \n",
    "    temp = pd.read_csv('../data/processed/epitope_calcs/Epitope_calc_'+protein+'.csv')\n",
    "    \n",
    "    if ind == 0:\n",
    "        df = temp\n",
    "    else: \n",
    "        df = df.append(temp)'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.000213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0       E  MYSFVSEE          0        8  0.002908      0.000213\n",
       "1       E  YSFVSEET          1        8  0.008297      0.000640\n",
       "2       E  SFVSEETG          2        8  0.008297      0.000640\n",
       "3       E  FVSEETGT          3        8  0.008297      0.000640\n",
       "4       E  VSEETGTL          4        8  0.008297      0.000640"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.000213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0       E  MYSFVSEE          0        8  0.002908      0.000213\n",
       "1       E  YSFVSEET          1        8  0.008297      0.000640\n",
       "2       E  SFVSEETG          2        8  0.008297      0.000640\n",
       "3       E  FVSEETGT          3        8  0.008297      0.000640\n",
       "4       E  VSEETGTL          4        8  0.008297      0.000640"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164786, 6)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21368</th>\n",
       "      <td>S</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.078112</td>\n",
       "      <td>0.007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21369</th>\n",
       "      <td>S</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.078112</td>\n",
       "      <td>0.007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21370</th>\n",
       "      <td>S</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.078112</td>\n",
       "      <td>0.007464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21371</th>\n",
       "      <td>S</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.075207</td>\n",
       "      <td>0.007251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21372</th>\n",
       "      <td>S</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.078112</td>\n",
       "      <td>0.007464</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "21368       S  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.078112   \n",
       "21369       S  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.078112   \n",
       "21370       S  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.078112   \n",
       "21371       S  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.075207   \n",
       "21372       S  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.078112   \n",
       "\n",
       "       perc_mutated  \n",
       "21368      0.007464  \n",
       "21369      0.007464  \n",
       "21370      0.007464  \n",
       "21371      0.007251  \n",
       "21372      0.007464  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop('index', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([156376.,   3915.,   1877.,    322.,    299.,    271.,      0.,\n",
       "           371.,   1081.,    274.]),\n",
       " array([0.        , 0.1064769 , 0.2129538 , 0.31943069, 0.42590759,\n",
       "        0.53238449, 0.63886139, 0.74533828, 0.85181518, 0.95829208,\n",
       "        1.06476898]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAD4CAYAAAAZ1BptAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAXaElEQVR4nO3df5Bd5X3f8fcnUsB2EixhLa4rqRWp5TSYScd4i5VmmjpWDAJnEH9AR0xTFFdTTQl20zRNDPVM6dhmBidpaZjBpIqlIjwuglI3aGKoqgFc2g5gFhMDglBtwIUNxFpbgrplDJH97R/3kXu9ukf7U3cl9H7N3LnnfJ/nnPs8s9J+9vy496aqkCRpkB9Z7AFIkk5choQkqZMhIUnqZEhIkjoZEpKkTksXewALbcWKFbVmzZrFHoYknVQee+yxb1XVyNT6my4k1qxZw9jY2GIPQ5JOKkn+16C6p5skSZ2mDYkkO5IcSPLUlPrHkzybZF+S3+6rX5tkvLVd2Fff0GrjSa7pq5+d5JEk+5PckeS0Vj+9rY+39jULMWFJ0szN5EjiVmBDfyHJLwAbgZ+pqvcCv9vq5wCbgPe2bT6XZEmSJcDNwEXAOcAVrS/AZ4Ebq2otcAjY0upbgENV9W7gxtZPkjRE04ZEVT0IHJxSvgq4oapeb30OtPpGYFdVvV5VzwPjwPntMV5Vz1XVG8AuYGOSAB8C7mrb7wQu7dvXzrZ8F7C+9ZckDclcr0m8B/jb7TTQf03yN1t9JfBiX7+JVuuqvwN4paoOT6n/0L5a+6ut/1GSbE0ylmRscnJyjlOSJE0115BYCiwH1gG/CdzZ/sof9Jd+zaHONG0/XKzaVlWjVTU6MnLUHVySpDmaa0hMAF+qnq8C3wdWtPrqvn6rgJeOUf8WsCzJ0il1+rdp7W/n6NNekqTjaK4h8Yf0riWQ5D3AafR+4e8GNrU7k84G1gJfBR4F1rY7mU6jd3F7d/U+p/wB4LK2383A3W15d1untd9ffq65JA3VtG+mS3I78EFgRZIJ4DpgB7Cj3Rb7BrC5/QLfl+RO4GngMHB1VX2v7edjwB5gCbCjqva1l/gEsCvJZ4DHge2tvh34QpJxekcQmxZgvpKkWcib7Y/z0dHRmus7rtdc8+UFHs3MfeOGjyzaa0tSkseqanRq3XdcS5I6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOk0bEkl2JDnQvqp0ats/S1JJVrT1JLkpyXiSJ5Kc19d3c5L97bG5r/7+JE+2bW5KklY/M8ne1n9vkuULM2VJ0kzN5EjiVmDD1GKS1cCHgRf6yhcBa9tjK3BL63smve/G/gBwPnBd3y/9W1rfI9sdea1rgPuqai1wX1uXJA3RtCFRVQ8CBwc03Qj8FtD/Jdkbgduq52FgWZJ3ARcCe6vqYFUdAvYCG1rbGVX1UPW+bPs24NK+fe1syzv76pKkIZnTNYkklwB/VlVfn9K0Enixb32i1Y5VnxhQB3hnVb0M0J7POsZ4tiYZSzI2OTk5hxlJkgaZdUgkeRvwSeBfDGoeUKs51GelqrZV1WhVjY6MjMx2c0lSh7kcSfw14Gzg60m+AawCvpbkL9E7Eljd13cV8NI09VUD6gDfbKejaM8H5jBWSdI8zDokqurJqjqrqtZU1Rp6v+jPq6o/B3YDV7a7nNYBr7ZTRXuAC5IsbxesLwD2tLbvJFnX7mq6Eri7vdRu4MhdUJv76pKkIZnJLbC3Aw8BP5VkIsmWY3S/B3gOGAf+APhVgKo6CHwaeLQ9PtVqAFcBn2/b/Clwb6vfAHw4yX56d1HdMLupSZLma+l0Harqimna1/QtF3B1R78dwI4B9THg3AH1bwPrpxufJOn48R3XkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKnTTL7jekeSA0me6qv9TpI/SfJEkv+UZFlf27VJxpM8m+TCvvqGVhtPck1f/ewkjyTZn+SOJKe1+ultfby1r1moSUuSZmYmRxK3Ahum1PYC51bVzwD/E7gWIMk5wCbgvW2bzyVZkmQJcDNwEXAOcEXrC/BZ4MaqWgscAra0+hbgUFW9G7ix9ZMkDdG0IVFVDwIHp9T+S1UdbqsPA6va8kZgV1W9XlXPA+PA+e0xXlXPVdUbwC5gY5IAHwLuatvvBC7t29fOtnwXsL71lyQNyUJck/gHwL1teSXwYl/bRKt11d8BvNIXOEfqP7Sv1v5q63+UJFuTjCUZm5ycnPeEJEk98wqJJJ8EDgNfPFIa0K3mUD/Wvo4uVm2rqtGqGh0ZGTn2oCVJM7Z0rhsm2Qz8ErC+qo788p4AVvd1WwW81JYH1b8FLEuytB0t9Pc/sq+JJEuBtzPltJck6fia05FEkg3AJ4BLquq1vqbdwKZ2Z9LZwFrgq8CjwNp2J9Np9C5u727h8gBwWdt+M3B33742t+XLgPv7wkiSNATTHkkkuR34ILAiyQRwHb27mU4H9rZryQ9X1T+qqn1J7gSepnca6uqq+l7bz8eAPcASYEdV7Wsv8QlgV5LPAI8D21t9O/CFJOP0jiA2LcB8JUmzMG1IVNUVA8rbB9SO9L8euH5A/R7gngH15+jd/TS1/l3g8unGJ0k6fnzHtSSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqdO0IZFkR5IDSZ7qq52ZZG+S/e15easnyU1JxpM8keS8vm02t/77k2zuq78/yZNtm5vSvg+16zUkScMzkyOJW4ENU2rXAPdV1VrgvrYOcBGwtj22ArdA7xc+ve/G/gC9ryq9ru+X/i2t75HtNkzzGpKkIZk2JKrqQeDglPJGYGdb3glc2le/rXoeBpYleRdwIbC3qg5W1SFgL7ChtZ1RVQ9VVQG3TdnXoNeQJA3JXK9JvLOqXgZoz2e1+krgxb5+E612rPrEgPqxXuMoSbYmGUsyNjk5OccpSZKmWugL1xlQqznUZ6WqtlXVaFWNjoyMzHZzSVKHuYbEN9upItrzgVafAFb39VsFvDRNfdWA+rFeQ5I0JHMNid3AkTuUNgN399WvbHc5rQNebaeK9gAXJFneLlhfAOxpbd9Jsq7d1XTllH0Neg1J0pAsna5DktuBDwIrkkzQu0vpBuDOJFuAF4DLW/d7gIuBceA14KMAVXUwyaeBR1u/T1XVkYvhV9G7g+qtwL3twTFeQ5I0JNOGRFVd0dG0fkDfAq7u2M8OYMeA+hhw7oD6twe9hiRpeHzHtSSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqNK+QSPLrSfYleSrJ7UnekuTsJI8k2Z/kjiSntb6nt/Xx1r6mbz/XtvqzSS7sq29otfEk18xnrJKk2ZtzSCRZCfxjYLSqzgWWAJuAzwI3VtVa4BCwpW2yBThUVe8Gbmz9SHJO2+69wAbgc0mWJFkC3AxcBJwDXNH6SpKGZL6nm5YCb02yFHgb8DLwIeCu1r4TuLQtb2zrtPb1SdLqu6rq9ap6HhgHzm+P8ap6rqreAHa1vpKkIZlzSFTVnwG/C7xALxxeBR4DXqmqw63bBLCyLa8EXmzbHm7939Ffn7JNV/0oSbYmGUsyNjk5OdcpSZKmmM/ppuX0/rI/G/jLwI/ROzU0VR3ZpKNttvWji1Xbqmq0qkZHRkamG7okaYbmc7rpF4Hnq2qyqv4C+BLwt4Bl7fQTwCrgpbY8AawGaO1vBw7216ds01WXJA3JfELiBWBdkre1awvrgaeBB4DLWp/NwN1teXdbp7XfX1XV6pva3U9nA2uBrwKPAmvb3VKn0bu4vXse45UkzdLS6bsMVlWPJLkL+BpwGHgc2AZ8GdiV5DOttr1tsh34QpJxekcQm9p+9iW5k17AHAaurqrvAST5GLCH3p1TO6pq31zHK0mavTmHBEBVXQdcN6X8HL07k6b2/S5wecd+rgeuH1C/B7hnPmOUJM2d77iWJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1mldIJFmW5K4kf5LkmSQ/m+TMJHuT7G/Py1vfJLkpyXiSJ5Kc17efza3//iSb++rvT/Jk2+am9l3akqQhme+RxO8B/7mq/jrwN4BngGuA+6pqLXBfWwe4CFjbHluBWwCSnEnvK1A/QO9rT687Eiytz9a+7TbMc7ySpFmYc0gkOQP4eWA7QFW9UVWvABuBna3bTuDStrwRuK16HgaWJXkXcCGwt6oOVtUhYC+wobWdUVUPVVUBt/XtS5I0BPM5kvhJYBL4d0keT/L5JD8GvLOqXgZoz2e1/iuBF/u2n2i1Y9UnBtSPkmRrkrEkY5OTk/OYkiSp33xCYilwHnBLVb0P+L/8/1NLgwy6nlBzqB9drNpWVaNVNToyMnLsUUuSZmw+ITEBTFTVI239Lnqh8c12qoj2fKCv/+q+7VcBL01TXzWgLkkakjmHRFX9OfBikp9qpfXA08Bu4MgdSpuBu9vybuDKdpfTOuDVdjpqD3BBkuXtgvUFwJ7W9p0k69pdTVf27UuSNARL57n9x4EvJjkNeA74KL3guTPJFuAF4PLW9x7gYmAceK31paoOJvk08Gjr96mqOtiWrwJuBd4K3NsekqQhmVdIVNUfA6MDmtYP6FvA1R372QHsGFAfA86dzxglSXPnO64lSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUqd5h0SSJUkeT/JHbf3sJI8k2Z/kjvb91yQ5va2Pt/Y1ffu4ttWfTXJhX31Dq40nuWa+Y5Ukzc5CHEn8GvBM3/pngRurai1wCNjS6luAQ1X1buDG1o8k5wCbgPcCG4DPteBZAtwMXAScA1zR+kqShmReIZFkFfAR4PNtPcCHgLtal53ApW15Y1unta9v/TcCu6rq9ap6HhgHzm+P8ap6rqreAHa1vpKkIZnvkcS/AX4L+H5bfwfwSlUdbusTwMq2vBJ4EaC1v9r6/6A+ZZuu+lGSbE0ylmRscnJynlOSJB0x55BI8kvAgap6rL88oGtN0zbb+tHFqm1VNVpVoyMjI8cYtSRpNpbOY9ufAy5JcjHwFuAMekcWy5IsbUcLq4CXWv8JYDUwkWQp8HbgYF/9iP5tuuqSpCGY85FEVV1bVauqag29C8/3V9XfAx4ALmvdNgN3t+XdbZ3Wfn9VVatvanc/nQ2sBb4KPAqsbXdLndZeY/dcxytJmr35HEl0+QSwK8lngMeB7a2+HfhCknF6RxCbAKpqX5I7gaeBw8DVVfU9gCQfA/YAS4AdVbXvOIxXktRhQUKiqr4CfKUtP0fvzqSpfb4LXN6x/fXA9QPq9wD3LMQYJUmz5zuuJUmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJneYcEklWJ3kgyTNJ9iX5tVY/M8neJPvb8/JWT5KbkowneSLJeX372tz670+yua/+/iRPtm1uSpL5TFaSNDvzOZI4DPxGVf00sA64Osk5wDXAfVW1FrivrQNcBKxtj63ALdALFeA64AP0vvb0uiPB0vps7dtuwzzGK0mapTmHRFW9XFVfa8vfAZ4BVgIbgZ2t207g0ra8Ebiteh4GliV5F3AhsLeqDlbVIWAvsKG1nVFVD1VVAbf17UuSNAQLck0iyRrgfcAjwDur6mXoBQlwVuu2Enixb7OJVjtWfWJAfdDrb00ylmRscnJyvtORJDXzDokkPw78R+CfVNX/PlbXAbWaQ/3oYtW2qhqtqtGRkZHphixJmqF5hUSSH6UXEF+sqi+18jfbqSLa84FWnwBW922+CnhpmvqqAXVJ0pDM5+6mANuBZ6rqX/c17QaO3KG0Gbi7r35lu8tpHfBqOx21B7ggyfJ2wfoCYE9r+06Sde21ruzblyRpCJbOY9ufA/4+8GSSP261fw7cANyZZAvwAnB5a7sHuBgYB14DPgpQVQeTfBp4tPX7VFUdbMtXAbcCbwXubQ9J0pDMOSSq6r8z+LoBwPoB/Qu4umNfO4AdA+pjwLlzHaMkaX58x7UkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkTvP5ZjotoDXXfHlRXvcbN3xkUV5X0snBIwlJUqcT/kgiyQbg94AlwOer6oZFHtKbikcwko7lhA6JJEuAm4EPAxPAo0l2V9XTizsySSc6/wBaGCd0SADnA+NV9RxAkl3ARsCQOMkt1n/gU9Vi/eI6FX/Oiznn4/FzPtFDYiXwYt/6BPCBqZ2SbAW2ttX/k+TZOb7eCuBbc9z2ZHEqzBFOjXnOeI757HEeyfHjz3EW5vlz/quDiid6SGRArY4qVG0Dts37xZKxqhqd735OZKfCHOHUmKdzfHM40ed4ot/dNAGs7ltfBby0SGORpFPOiR4SjwJrk5yd5DRgE7B7kcckSaeME/p0U1UdTvIxYA+9W2B3VNW+4/iS8z5ldRI4FeYIp8Y8neObwwk9x1QddYpfkiTgxD/dJElaRIaEJKnTKRkSSTYkeTbJeJJrBrSfnuSO1v5IkjXDH+X8zGCO/zTJ00meSHJfkoH3SJ/IpptjX7/LklSSE/Y2wy4zmWOSv9t+lvuS/Pthj3EhzODf619J8kCSx9u/2YsXY5xzlWRHkgNJnupoT5Kb2vyfSHLesMfYqapOqQe9C+B/CvwkcBrwdeCcKX1+Ffj9trwJuGOxx30c5vgLwNva8lVvxjm2fj8BPAg8DIwu9riPw89xLfA4sLytn7XY4z5O89wGXNWWzwG+sdjjnuUcfx44D3iqo/1i4F567w1bBzyy2GM+8jgVjyR+8FEfVfUGcOSjPvptBHa25buA9UkGvbHvRDXtHKvqgap6ra0+TO89KCeTmfwcAT4N/Dbw3WEOboHMZI7/ELi5qg4BVNWBIY9xIcxkngWc0Zbfzkn2fqmqehA4eIwuG4HbqudhYFmSdw1ndMd2KobEoI/6WNnVp6oOA68C7xjK6BbGTObYbwu9v2JOJtPOMcn7gNVV9UfDHNgCmsnP8T3Ae5L8jyQPt09NPtnMZJ7/EvjlJBPAPcDHhzO0oZnt/9mhOaHfJ3GczOSjPmb0cSAnsBmPP8kvA6PA3zmuI1p4x5xjkh8BbgR+ZVgDOg5m8nNcSu+U0wfpHQ3+tyTnVtUrx3lsC2km87wCuLWq/lWSnwW+0Ob5/eM/vKE4YX/nnIpHEjP5qI8f9EmylN7h7bEOFU80M/o4kyS/CHwSuKSqXh/S2BbKdHP8CeBc4CtJvkHvPO/uk+zi9Uz/rd5dVX9RVc8Dz9ILjZPJTOa5BbgToKoeAt5C74Px3ixO2I8gOhVDYiYf9bEb2NyWLwPur3Z16SQx7RzbqZh/Sy8gTsbz2MecY1W9WlUrqmpNVa2hd93lkqoaW5zhzslM/q3+Ib2bEEiygt7pp+eGOsr5m8k8XwDWAyT5aXohMTnUUR5fu4Er211O64BXq+rlxR4UnIKnm6rjoz6SfAoYq6rdwHZ6h7Pj9I4gNi3eiGdvhnP8HeDHgf/Qrsm/UFWXLNqgZ2mGczypzXCOe4ALkjwNfA/4zar69uKNevZmOM/fAP4gya/TOw3zKyfTH25Jbqd3SnBFu65yHfCjAFX1+/Sus1wMjAOvAR9dnJEezY/lkCR1OhVPN0mSZsiQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmd/h+EI6eAu12y2wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(df.entropy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10169"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.perc_mutated==0.0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164786, 6)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([162490.,    570.,    328.,    543.,    285.,      0.,      0.,\n",
       "             0.,      0.,    570.]),\n",
       " array([0.        , 0.0689486 , 0.13789721, 0.20684581, 0.27579441,\n",
       "        0.34474302, 0.41369162, 0.48264022, 0.55158882, 0.62053743,\n",
       "        0.68948603]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAD4CAYAAAAZ1BptAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAXuUlEQVR4nO3dfZBd9X3f8fcnUsEPDeZpcV1JU8nx2qlg3DEottpMM46VgLBdi06hFdMExaXVhAg7bdPaUHdKxzZTnGRCzRTTIUZFeFxkqrpFjUVUBXA96ZiHtbHBgmCtBYUNxKwtQdx4DBH+9o/7k3O9ukf7JN1dqvdrZmfP+Z7fOfd776zuR+fh3pOqQpKkQX5ioRuQJC1ehoQkqZMhIUnqZEhIkjoZEpKkTksXuoFj7cwzz6yVK1cudBuS9Iryla985TtVNTK1/v9dSKxcuZKxsbGFbkOSXlGS/J9BdQ83SZI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjpN+4nrJFuB9wLPVdU5ffUPAFcCh4AvVNWHWv1q4HLgZeCDVbW71dcDnwSWAJ+uqutafRWwHTgd+Crwy1X1UpKTgduA84DvAv+gqp48Fk+6y8qrvnA8N39UT173ngV7bEnqMpM9iVuB9f2FJD8PbADeWlVnA7/d6quBjcDZbZ1PJVmSZAlwI3AhsBq4tI0F+ARwfVWNAgfpBQzt98GqehNwfRsnSRqiaUOiqr4EHJhSvgK4rqpebGOea/UNwPaqerGqngDGgbe3n/Gq2l9VL9Hbc9iQJMC7gB1t/W3ARX3b2tamdwDr2nhJ0pDM9ZzEm4G/neT+JP8ryc+0+jLg6b5xE63WVT8DeL6qDk2p/9i22vIX2vgjJNmcZCzJ2OTk5ByfkiRpqrmGxFLgNGAt8C+BO9r/8gf9T7/mUGeaZT9erLq5qtZU1ZqRkSO+6VaSNEdzDYkJ4PPV8wDwQ+DMVl/RN2458MxR6t8BTk2ydEqd/nXa8tdx5GEvSdJxNNeQ+O/0ziWQ5M3ASfTe8HcCG5Oc3K5aGgUeAB4ERpOsSnISvZPbO6uqgHuBi9t2NwF3tumdbZ62/J42XpI0JDO5BPZ24J3AmUkmgGuArcDWJN8AXgI2tTfwvUnuAB6ld2nslqp6uW3nSmA3vUtgt1bV3vYQHwa2J/k48BBwS6vfAnwmyTi9PYiNx+D5SpJmYdqQqKpLOxb9Usf4a4FrB9R3AbsG1PfTu/ppav0HwCXT9SdJOn78xLUkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkTtOGRJKtSZ5rd6GbuuxfJKkkZ7b5JLkhyXiSh5Oc2zd2U5J97WdTX/28JI+0dW5IklY/PcmeNn5PktOOzVOWJM3UTPYkbgXWTy0mWQH8IvBUX/lCeve1HgU2Aze1safTu+3pO+jdhe6avjf9m9rYw+sdfqyrgLurahS4u81LkoZo2pCoqi/Ru8f0VNcDHwKqr7YBuK167gNOTfIG4AJgT1UdqKqDwB5gfVt2SlV9ud0j+zbgor5tbWvT2/rqkqQhmdM5iSTvA/64qr4+ZdEy4Om++YlWO1p9YkAd4PVV9SxA+33WUfrZnGQsydjk5OQcnpEkaZBZh0SS1wAfAf7NoMUDajWH+qxU1c1Vtaaq1oyMjMx2dUlSh7nsSfwUsAr4epIngeXAV5P8FXp7Aiv6xi4HnpmmvnxAHeDb7XAU7fdzc+hVkjQPsw6Jqnqkqs6qqpVVtZLeG/25VfUnwE7gsnaV01rghXaoaDdwfpLT2gnr84Hdbdn3kqxtVzVdBtzZHmoncPgqqE19dUnSkMzkEtjbgS8Db0kykeTyowzfBewHxoHfBX4NoKoOAB8DHmw/H201gCuAT7d1vgXc1erXAb+YZB+9q6ium91TkyTN19LpBlTVpdMsX9k3XcCWjnFbga0D6mPAOQPq3wXWTdefJOn48RPXkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkTjO5fenWJM8l+UZf7beS/FGSh5P8tySn9i27Osl4kseTXNBXX99q40mu6quvSnJ/kn1JPpfkpFY/uc2Pt+Urj9WTliTNzEz2JG4F1k+p7QHOqaq3At8ErgZIshrYCJzd1vlUkiVJlgA3AhcCq4FL21iATwDXV9UocBA4fA/ty4GDVfUm4Po2TpI0RNOGRFV9CTgwpfY/q+pQm70PWN6mNwDbq+rFqnoCGAfe3n7Gq2p/Vb0EbAc2JAnwLmBHW38bcFHftra16R3AujZekjQkx+KcxD8C7mrTy4Cn+5ZNtFpX/Qzg+b7AOVz/sW215S+08UdIsjnJWJKxycnJeT8hSVLPvEIiyUeAQ8BnD5cGDKs51I+2rSOLVTdX1ZqqWjMyMnL0piVJM7Z0rism2QS8F1hXVYffvCeAFX3DlgPPtOlB9e8ApyZZ2vYW+scf3tZEkqXA65hy2EuSdHzNaU8iyXrgw8D7qur7fYt2AhvblUmrgFHgAeBBYLRdyXQSvZPbO1u43Atc3NbfBNzZt61Nbfpi4J6+MJIkDcG0exJJbgfeCZyZZAK4ht7VTCcDe9q55Puq6leram+SO4BH6R2G2lJVL7ftXAnsBpYAW6tqb3uIDwPbk3wceAi4pdVvAT6TZJzeHsTGY/B8JUmzMG1IVNWlA8q3DKgdHn8tcO2A+i5g14D6fnpXP02t/wC4ZLr+JEnHj5+4liR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktRp2pBIsjXJc0m+0Vc7PcmeJPva79NaPUluSDKe5OEk5/ats6mN39fuj324fl6SR9o6N6Td6q7rMSRJwzOTPYlbgfVTalcBd1fVKHB3mwe4kN59rUeBzcBN0HvDp3fb03fQuwvdNX1v+je1sYfXWz/NY0iShmTakKiqL9G7x3S/DcC2Nr0NuKivflv13AecmuQNwAXAnqo6UFUHgT3A+rbslKr6clUVcNuUbQ16DEnSkMz1nMTrq+pZgPb7rFZfBjzdN26i1Y5WnxhQP9pjHCHJ5iRjScYmJyfn+JQkSVMd6xPXGVCrOdRnpapurqo1VbVmZGRktqtLkjrMNSS+3Q4V0X4/1+oTwIq+ccuBZ6apLx9QP9pjSJKGZK4hsRM4fIXSJuDOvvpl7SqntcAL7VDRbuD8JKe1E9bnA7vbsu8lWduuarpsyrYGPYYkaUiWTjcgye3AO4Ezk0zQu0rpOuCOJJcDTwGXtOG7gHcD48D3gfcDVNWBJB8DHmzjPlpVh0+GX0HvCqpXA3e1H47yGJKkIZk2JKrq0o5F6waMLWBLx3a2AlsH1MeAcwbUvzvoMSRJw+MnriVJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ3mFRJJ/lmSvUm+keT2JK9KsirJ/Un2JflckpPa2JPb/HhbvrJvO1e3+uNJLuirr2+18SRXzadXSdLszTkkkiwDPgisqapzgCXARuATwPVVNQocBC5vq1wOHKyqNwHXt3EkWd3WOxtYD3wqyZIkS4AbgQuB1cClbawkaUjme7hpKfDqJEuB1wDPAu8CdrTl24CL2vSGNk9bvi5JWn17Vb1YVU8A48Db2894Ve2vqpeA7W2sJGlI5hwSVfXHwG8DT9ELhxeArwDPV9WhNmwCWNamlwFPt3UPtfFn9NenrNNVP0KSzUnGkoxNTk7O9SlJkqaYz+Gm0+j9z34V8FeB19I7NDRVHV6lY9ls60cWq26uqjVVtWZkZGS61iVJMzSfw02/ADxRVZNV9efA54G/BZzaDj8BLAeeadMTwAqAtvx1wIH++pR1uuqSpCGZT0g8BaxN8pp2bmEd8ChwL3BxG7MJuLNN72zztOX3VFW1+sZ29dMqYBR4AHgQGG1XS51E7+T2znn0K0mapaXTDxmsqu5PsgP4KnAIeAi4GfgCsD3Jx1vtlrbKLcBnkozT24PY2LazN8kd9ALmELClql4GSHIlsJvelVNbq2rvXPuVJM3enEMCoKquAa6ZUt5P78qkqWN/AFzSsZ1rgWsH1HcBu+bToyRp7vzEtSSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqRO8wqJJKcm2ZHkj5I8luRvJjk9yZ4k+9rv09rYJLkhyXiSh5Oc27edTW38viSb+urnJXmkrXNDu02qJGlI5rsn8Ung96vqp4G/ATwGXAXcXVWjwN1tHuBCevevHgU2AzcBJDmd3t3t3kHvjnbXHA6WNmZz33rr59mvJGkW5hwSSU4Bfo52D+uqeqmqngc2ANvasG3ARW16A3Bb9dwHnJrkDcAFwJ6qOlBVB4E9wPq27JSq+nJVFXBb37YkSUMwnz2JNwKTwH9K8lCSTyd5LfD6qnoWoP0+q41fBjzdt/5Eqx2tPjGgfoQkm5OMJRmbnJycx1OSJPWbT0gsBc4FbqqqtwF/xl8cWhpk0PmEmkP9yGLVzVW1pqrWjIyMHL1rSdKMzSckJoCJqrq/ze+gFxrfboeKaL+f6xu/om/95cAz09SXD6hLkoZkziFRVX8CPJ3kLa20DngU2AkcvkJpE3Bnm94JXNaucloLvNAOR+0Gzk9yWjthfT6wuy37XpK17aqmy/q2JUkagqXzXP8DwGeTnATsB95PL3juSHI58BRwSRu7C3g3MA58v42lqg4k+RjwYBv30ao60KavAG4FXg3c1X4kSUMyr5Coqq8BawYsWjdgbAFbOrazFdg6oD4GnDOfHiVJc+cnriVJnQwJSVInQ0KS1MmQkCR1MiQkSZ0MCUlSJ0NCktTJkJAkdTIkJEmdDAlJUidDQpLUyZCQJHUyJCRJnQwJSVInQ0KS1MmQkCR1MiQkSZ3mHRJJliR5KMnvtflVSe5Psi/J59qtTUlycpsfb8tX9m3j6lZ/PMkFffX1rTae5Kr59ipJmp1jsSfx68BjffOfAK6vqlHgIHB5q18OHKyqNwHXt3EkWQ1sBM4G1gOfasGzBLgRuBBYDVzaxkqShmReIZFkOfAe4NNtPsC7gB1tyDbgoja9oc3Tlq9r4zcA26vqxap6AhgH3t5+xqtqf1W9BGxvYyVJQzLfPYl/D3wI+GGbPwN4vqoOtfkJYFmbXgY8DdCWv9DG/6g+ZZ2u+hGSbE4ylmRscnJynk9JknTYnEMiyXuB56rqK/3lAUNrmmWzrR9ZrLq5qtZU1ZqRkZGjdC1Jmo2l81j3Z4H3JXk38CrgFHp7FqcmWdr2FpYDz7TxE8AKYCLJUuB1wIG++mH963TVJUlDMOc9iaq6uqqWV9VKeiee76mqfwjcC1zchm0C7mzTO9s8bfk9VVWtvrFd/bQKGAUeAB4ERtvVUie1x9g5134lSbM3nz2JLh8Gtif5OPAQcEur3wJ8Jsk4vT2IjQBVtTfJHcCjwCFgS1W9DJDkSmA3sATYWlV7j0O/kqQOxyQkquqLwBfb9H56VyZNHfMD4JKO9a8Frh1Q3wXsOhY9SpJmz09cS5I6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSeo055BIsiLJvUkeS7I3ya+3+ulJ9iTZ136f1upJckOS8SQPJzm3b1ub2vh9STb11c9L8khb54Ykmc+TlSTNznz2JA4Bv1FVfx1YC2xJshq4Cri7qkaBu9s8wIX07l89CmwGboJeqADXAO+gd0e7aw4HSxuzuW+99fPoV5I0S3MOiap6tqq+2qa/BzwGLAM2ANvasG3ARW16A3Bb9dwHnJrkDcAFwJ6qOlBVB4E9wPq27JSq+nJVFXBb37YkSUNwTM5JJFkJvA24H3h9VT0LvSABzmrDlgFP96020WpHq08MqA96/M1JxpKMTU5OzvfpSJKaeYdEkr8M/Ffgn1bVnx5t6IBazaF+ZLHq5qpaU1VrRkZGpmtZkjRD8wqJJH+JXkB8tqo+38rfboeKaL+fa/UJYEXf6suBZ6apLx9QlyQNyXyubgpwC/BYVf1O36KdwOErlDYBd/bVL2tXOa0FXmiHo3YD5yc5rZ2wPh/Y3ZZ9L8na9liX9W1LkjQES+ex7s8Cvww8kuRrrfavgOuAO5JcDjwFXNKW7QLeDYwD3wfeD1BVB5J8DHiwjftoVR1o01cAtwKvBu5qP5KkIZlzSFTVHzL4vAHAugHjC9jSsa2twNYB9THgnLn2KEmaHz9xLUnqZEhIkjoZEpKkToaEJKmTISFJ6mRISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqROhoQkqZMhIUnqZEhIkjoZEpKkToaEJKmTISFJ6rToQyLJ+iSPJxlPctVC9yNJJ5JFHRJJlgA3AhcCq4FLk6xe2K4k6cQx53tcD8nbgfGq2g+QZDuwAXh0Qbs6DlZe9YWFbkHH2ZPXvWehW9AQLOS/5ePxN7bYQ2IZ8HTf/ATwjqmDkmwGNrfZ/5vk8Tk+3pnAd+a47kKw3+PvmPWcTxyLrUzrlfYa2+8xNOBvbDb9/rVBxcUeEhlQqyMKVTcDN8/7wZKxqloz3+0Mi/0ef6+0nu33+DoR+13U5yTo7Tms6JtfDjyzQL1I0glnsYfEg8BoklVJTgI2AjsXuCdJOmEs6sNNVXUoyZXAbmAJsLWq9h7Hh5z3Iashs9/j75XWs/0eXydcv6k64hC/JEnA4j/cJElaQIaEJKnTCRkS033VR5KTk3yuLb8/ycrhd/lj/UzX788l+WqSQ0kuXogep/QzXb//PMmjSR5OcneSgddnD8sM+v3VJI8k+VqSP1zoT/3P9KtqklycpJIs+CWbM3iNfyXJZHuNv5bkHy9En339TPsaJ/n77e94b5L/POwep/Qy3et7fd9r+80kz89441V1Qv3QOwH+LeCNwEnA14HVU8b8GvAf2/RG4HOLvN+VwFuB24CLXwGv788Dr2nTV7wCXt9T+qbfB/z+Yu63jftJ4EvAfcCaV8DfxK8A/2Eh+5xlv6PAQ8Bpbf6sxdzvlPEfoHcR0Iy2fyLuSfzoqz6q6iXg8Fd99NsAbGvTO4B1SQZ9sG8Ypu23qp6sqoeBHy5Eg1PMpN97q+r7bfY+ep9/WSgz6fdP+2Zfy4APdA7RTP5+AT4G/Cbwg2E212GmPS8WM+n3nwA3VtVBgKp6bsg99pvt63spcPtMN34ihsSgr/pY1jWmqg4BLwBnDKW7I82k38Vktv1eDtx1XDs6uhn1m2RLkm/Re+P94JB6G2TafpO8DVhRVb83zMaOYqZ/E3+vHYLckWTFgOXDMpN+3wy8Ocn/TnJfkvVD6+5IM/431w7trgLumenGT8SQmMlXfczo60CGZDH1MhMz7jfJLwFrgN86rh0d3Uy/+uXGqvop4MPAvz7uXXU7ar9JfgK4HviNoXU0vZm8xv8DWFlVbwX+gL/Yk18IM+l3Kb1DTu+k9z/zTyc59Tj31WU27xEbgR1V9fJMN34ihsRMvurjR2OSLAVeBxwYSndHeqV9NcmM+k3yC8BHgPdV1YtD6m2Q2b6+24GLjmtHRzddvz8JnAN8McmTwFpg5wKfvJ72Na6q7/b9HfwucN6Qehtkpu8Rd1bVn1fVE8Dj9EJjIczmb3gjszjUBJyQJ66XAvvp7XIdPslz9pQxW/jxE9d3LOZ++8beysKfuJ7J6/s2eifaRl8hfw+jfdN/BxhbzP1OGf9FFv7E9Uxe4zf0Tf9d4L5F3u96YFubPpPe4Z4zFmu/bdxbgCdpH6Ke8fYX8o9nAf8I3g18s71RfaTVPkrvf7UArwL+CzAOPAC8cZH3+zP0/jfxZ8B3gb2LvN8/AL4NfK397Fzk/X4S2Nt6vfdob8qLod8pYxc8JGb4Gv+79hp/vb3GP73I+w3wO/TubfMIsHEx99vm/y1w3Wy37ddySJI6nYjnJCRJM2RISJI6GRKSpE6GhCSpkyEhSepkSEiSOhkSkqRO/w9hf9vYeAaxEgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(df.perc_mutated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "70623"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.perc_mutated <0.001).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10169"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.entropy == 0.0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10169"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.perc_mutated==0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002908</td>\n",
       "      <td>0.000213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008297</td>\n",
       "      <td>0.000640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0       E  MYSFVSEE          0        8  0.002908      0.000213\n",
       "1       E  YSFVSEET          1        8  0.008297      0.000640\n",
       "2       E  SFVSEETG          2        8  0.008297      0.000640\n",
       "3       E  FVSEETGT          3        8  0.008297      0.000640\n",
       "4       E  VSEETGTL          4        8  0.008297      0.000640"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../data/processed/Hu1_epitope_entropies.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Looking at and flagging the masked positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>masked_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00288</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len  entropy  perc_mutated  \\\n",
       "0       E  MYSFVSEE          0        8  0.00288      0.000211   \n",
       "1       E  YSFVSEET          1        8  0.00576      0.000422   \n",
       "2       E  SFVSEETG          2        8  0.00576      0.000422   \n",
       "3       E  FVSEETGT          3        8  0.00576      0.000422   \n",
       "4       E  VSEETGTL          4        8  0.00576      0.000422   \n",
       "\n",
       "   masked_position  \n",
       "0                0  \n",
       "1                0  \n",
       "2                0  \n",
       "3                0  \n",
       "4                0  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data/processed/epitope_entropies.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "852"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.masked_position.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein\n",
       "E          0\n",
       "M          0\n",
       "N          0\n",
       "ORF10      0\n",
       "ORF14      0\n",
       "ORF1a    282\n",
       "ORF1b    285\n",
       "ORF3a      0\n",
       "ORF6       0\n",
       "ORF7a      0\n",
       "ORF7b      0\n",
       "ORF8       0\n",
       "ORF9b      0\n",
       "S        285\n",
       "Name: masked_position, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('protein').masked_position.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein  epi_len\n",
       "ORF1a    8           8\n",
       "         9           9\n",
       "         10         10\n",
       "         11         11\n",
       "         13         13\n",
       "         14         14\n",
       "         15         15\n",
       "         16         16\n",
       "         17         17\n",
       "         18         18\n",
       "         19         19\n",
       "         20         20\n",
       "         21         21\n",
       "         22         22\n",
       "         23         23\n",
       "         24         23\n",
       "         25         23\n",
       "Name: masked_position, dtype: int64"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.protein=='ORF1a'].groupby(['protein', 'epi_len']).masked_position.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 8,  9, 10, 11, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25])"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.epi_len.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "285"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.arange(8,26).sum()-12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to mask out all of these positions by setting them to 100% prob of mutating. and have \n",
    "#anything within them be set to prob of 999. \n",
    "# this will help flag them as being unique. \n",
    "# actually just create a new masked column feature. \n",
    "\n",
    "masked_nt_positions = [18529, 29849, 29851, 29853, 13402, 24389, 24390]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ORF1a': '266..13468',\n",
       " 'ORF1b': '13468..21555',\n",
       " 'S': '21563..25384',\n",
       " 'ORF3a': '25393..26220',\n",
       " 'E': '26245..26472',\n",
       " 'M': '26523..27191',\n",
       " 'ORF6': '27202..27387',\n",
       " 'ORF7a': '27394..27759',\n",
       " 'ORF7b': '27756..27887',\n",
       " 'ORF8': '27894..28259',\n",
       " 'N': '28274..29533',\n",
       " 'ORF9b': '28284..28577',\n",
       " 'ORF10': '29558..29674',\n",
       " 'ORF14': '28734..28955'}"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting the nt regions\n",
    "\n",
    "with open('nextstrain_covid19_ref_protein_pos.txt', 'r') as f: \n",
    "    lines = f.readlines()\n",
    "take_next_line = False\n",
    "regions = []\n",
    "proteins = []\n",
    "for l in lines:\n",
    "    if 'CDS ' in l:\n",
    "        regions.append(l.split('CDS')[1].strip())\n",
    "        take_next_line=True\n",
    "        \n",
    "    elif take_next_line:\n",
    "        proteins.append(l.split('gene=')[1].strip().strip('\"'))\n",
    "        take_next_line=False\n",
    "        \n",
    "protein_regions = {p:r for p,r in zip(proteins, regions)}\n",
    "protein_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_proteins = [18529-13468, \n",
    "               13402-266,\n",
    "               24389-21563, \n",
    "               24390-21563]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "4400-4378"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_proteins = np.floor(np.asarray(in_proteins)/3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1687., 4378.,  942.,  942.])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "the_proteins = ['ORF1b', 'ORF1a', 'S', 'S']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>masked_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00288</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len  entropy  perc_mutated  \\\n",
       "0       E  MYSFVSEE          0        8  0.00288      0.000211   \n",
       "1       E  YSFVSEET          1        8  0.00576      0.000422   \n",
       "2       E  SFVSEETG          2        8  0.00576      0.000422   \n",
       "3       E  FVSEETGT          3        8  0.00576      0.000422   \n",
       "4       E  VSEETGTL          4        8  0.00576      0.000422   \n",
       "\n",
       "   masked_position  \n",
       "0                0  \n",
       "1                0  \n",
       "2                0  \n",
       "3                0  \n",
       "4                0  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['masked_position'] = 0\n",
    "for pos, protein in zip(in_proteins, the_proteins):\n",
    "\n",
    "    protein_mask = protein==df.protein # select relevant protein\n",
    "    seq_start = df['start_pos']\n",
    "    seq_end = df['start_pos']+(df['epi_len']-1)\n",
    "    \n",
    "    in_region = np.logical_and(pos >= seq_start,pos <= seq_end)\n",
    "    #if all 4 matter::: \n",
    "    #.    front_in_region = np.logical_and(glyco_start >= seq_start,glyco_start <= seq_end)\n",
    "    #     end_in_region = np.logical_and(glyco_end >= seq_start,glyco_end <= seq_end)\n",
    "    #.    in_region = np.logical_or(front_in_region, end_in_region)\n",
    "    in_region_and_protein = np.logical_and(protein_mask,in_region)\n",
    "    \n",
    "    df.loc[in_region_and_protein,'masked_position'] = 1\n",
    "    #apply protein mask and then epitope mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>masked_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>88954</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>RIKIVQML</td>\n",
       "      <td>1686</td>\n",
       "      <td>8</td>\n",
       "      <td>0.005338</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein   epitope  start_pos  epi_len   entropy  perc_mutated  \\\n",
       "88954   ORF1b  RIKIVQML       1686        8  0.005338      0.000422   \n",
       "\n",
       "       masked_position  \n",
       "88954                1  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[np.logical_and(df.protein=='ORF1b', np.logical_and(df.start_pos==1686,df.epi_len==8 ))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>masked_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00288</td>\n",
       "      <td>0.000211</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.00576</td>\n",
       "      <td>0.000422</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len  entropy  perc_mutated  \\\n",
       "0       E  MYSFVSEE          0        8  0.00288      0.000211   \n",
       "1       E  YSFVSEET          1        8  0.00576      0.000422   \n",
       "2       E  SFVSEETG          2        8  0.00576      0.000422   \n",
       "3       E  FVSEETGT          3        8  0.00576      0.000422   \n",
       "4       E  VSEETGTL          4        8  0.00576      0.000422   \n",
       "\n",
       "   masked_position  \n",
       "0                0  \n",
       "1                0  \n",
       "2                0  \n",
       "3                0  \n",
       "4                0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164781</th>\n",
       "      <td>S</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.075762</td>\n",
       "      <td>0.007173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164782</th>\n",
       "      <td>S</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.075762</td>\n",
       "      <td>0.007173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164783</th>\n",
       "      <td>S</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.075762</td>\n",
       "      <td>0.007173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164784</th>\n",
       "      <td>S</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.072884</td>\n",
       "      <td>0.006962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164785</th>\n",
       "      <td>S</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.075762</td>\n",
       "      <td>0.007173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "164781       S  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.075762   \n",
       "164782       S  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.075762   \n",
       "164783       S  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.075762   \n",
       "164784       S  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.072884   \n",
       "164785       S  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.075762   \n",
       "\n",
       "        perc_mutated  \n",
       "164781      0.007173  \n",
       "164782      0.007173  \n",
       "164783      0.007173  \n",
       "164784      0.006962  \n",
       "164785      0.007173  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.protein.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164786, 6)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.epi_len.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.005363</td>\n",
       "      <td>0.000390</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  protein   epitope  start_pos  epi_len   entropy  perc_mutated  glyco_probs  \\\n",
       "0       E  MYSFVSEE          0        8  0.005363      0.000390          0.0   \n",
       "1       E  YSFVSEET          1        8  0.008044      0.000584          0.0   \n",
       "2       E  SFVSEETG          2        8  0.008044      0.000584          0.0   \n",
       "3       E  FVSEETGT          3        8  0.008044      0.000584          0.0   \n",
       "4       E  VSEETGTL          4        8  0.008044      0.000584          0.0   \n",
       "\n",
       "   crosses_cleavage  \n",
       "0                 0  \n",
       "1                 0  \n",
       "2                 0  \n",
       "3                 0  \n",
       "4                 0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.read_csv('../data/processed/pre_cleaning_AllEpitopeFeatures.csv')\n",
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>164764</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.073004</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164765</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.073004</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164766</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.073004</td>\n",
       "      <td>0.006817</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164767</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.068034</td>\n",
       "      <td>0.006428</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164768</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.070714</td>\n",
       "      <td>0.006623</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "164764      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.073004   \n",
       "164765      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.073004   \n",
       "164766      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.073004   \n",
       "164767      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.068034   \n",
       "164768      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.070714   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "164764      0.006817          0.0                 0  \n",
       "164765      0.006817          0.0                 0  \n",
       "164766      0.006817          0.0                 0  \n",
       "164767      0.006428          0.0                 0  \n",
       "164768      0.006623          0.0                 0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(164769, 8)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.005363</td>\n",
       "      <td>0.000390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0      0       E  MYSFVSEE          0        8  0.005363      0.000390\n",
       "1      1       E  YSFVSEET          1        8  0.008044      0.000584\n",
       "2      2       E  SFVSEETG          2        8  0.008044      0.000584\n",
       "3      3       E  FVSEETGT          3        8  0.008044      0.000584\n",
       "4      4       E  VSEETGTL          4        8  0.008044      0.000584"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run for a single example. \n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.protein.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>1002</td>\n",
       "      <td>E</td>\n",
       "      <td>VNVSLVKPSFYVYSRVKNLNSSRVP</td>\n",
       "      <td>46</td>\n",
       "      <td>25</td>\n",
       "      <td>0.055131</td>\n",
       "      <td>0.004480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>1003</td>\n",
       "      <td>E</td>\n",
       "      <td>NVSLVKPSFYVYSRVKNLNSSRVPD</td>\n",
       "      <td>47</td>\n",
       "      <td>25</td>\n",
       "      <td>0.064925</td>\n",
       "      <td>0.005259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1004</th>\n",
       "      <td>1004</td>\n",
       "      <td>E</td>\n",
       "      <td>VSLVKPSFYVYSRVKNLNSSRVPDL</td>\n",
       "      <td>48</td>\n",
       "      <td>25</td>\n",
       "      <td>0.081422</td>\n",
       "      <td>0.006623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1005</th>\n",
       "      <td>1005</td>\n",
       "      <td>E</td>\n",
       "      <td>SLVKPSFYVYSRVKNLNSSRVPDLL</td>\n",
       "      <td>49</td>\n",
       "      <td>25</td>\n",
       "      <td>0.081422</td>\n",
       "      <td>0.006623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1006</th>\n",
       "      <td>1006</td>\n",
       "      <td>E</td>\n",
       "      <td>LVKPSFYVYSRVKNLNSSRVPDLLV</td>\n",
       "      <td>50</td>\n",
       "      <td>25</td>\n",
       "      <td>0.106883</td>\n",
       "      <td>0.009155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "1002   1002       E  VNVSLVKPSFYVYSRVKNLNSSRVP         46       25  0.055131   \n",
       "1003   1003       E  NVSLVKPSFYVYSRVKNLNSSRVPD         47       25  0.064925   \n",
       "1004   1004       E  VSLVKPSFYVYSRVKNLNSSRVPDL         48       25  0.081422   \n",
       "1005   1005       E  SLVKPSFYVYSRVKNLNSSRVPDLL         49       25  0.081422   \n",
       "1006   1006       E  LVKPSFYVYSRVKNLNSSRVPDLLV         50       25  0.106883   \n",
       "\n",
       "      perc_mutated  \n",
       "1002      0.004480  \n",
       "1003      0.005259  \n",
       "1004      0.006623  \n",
       "1005      0.006623  \n",
       "1006      0.009155  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.protein=='E'].tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein: E\n",
      "window size 8\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def entropy_calc(x):\n",
    "        # where x is a list of numbers. \n",
    "        summ = 0\n",
    "        nc = np.sum(x)\n",
    "        for e in x: \n",
    "            p = e/nc\n",
    "            summ += p*np.log2(p)\n",
    "        return -summ\n",
    "    \n",
    "window_sizes = list(range(8,12)) + list(range(13,26))\n",
    "\n",
    "aa_file = amino_acid_files[0]\n",
    "protein = aa_file.split('_')[-1].split('.')[0]\n",
    "print('protein:', protein)\n",
    "with open(aa_file, \"rt\") as handle:\n",
    "    records = list(SeqIO.parse(handle, \"fasta\"))\n",
    "\n",
    "# getting rid of the node sequences!\n",
    "no_nodes = []\n",
    "for r in records: \n",
    "    if 'NODE_' not in r.id:\n",
    "        no_nodes.append(r)\n",
    "records = no_nodes\n",
    "\n",
    "# getting the reference sequence\n",
    "for ind, r in enumerate(records): \n",
    "    if r.id == 'Wuhan/IPBCAMS-WH-01/2019':#'Wuhan-Hu-1/2019':#'Wuhan/WH01/2019':#'Wuhan/IPBCAMS-WH-01/2019':\n",
    "        ref_seq_ind = ind\n",
    "\n",
    "ref_seq = str(records[ref_seq_ind].seq)[:-1]\n",
    "\n",
    "seqs = np.array(records)[:, :-1] # ignore the stop code at the end. \n",
    "\n",
    "for window in window_sizes:\n",
    "    print('window size', window)\n",
    "\n",
    "    window_res = []\n",
    "\n",
    "    # get epitope based column slices. gives a list of epi columns. \n",
    "    epi_columns = [seqs[:,i:i+window] for i in range(seqs.shape[1]-window+1)]\n",
    "\n",
    "    for col_ind, col in enumerate(epi_columns):\n",
    "        # count all of the unique epitopes here. \n",
    "\n",
    "        #first need to convert each of the columns into strings:\n",
    "        col = np.asarray(  [  ''.join(col[i,:]) for i in range(col.shape[0]) ])\n",
    "        unique, counts = np.unique(col, return_counts=True) \n",
    "        ent = entropy_calc(counts)\n",
    "\n",
    "        # useful for percentages\n",
    "        ref_epitope = col[ref_seq_ind]\n",
    "        count_dict = dict(zip(unique, counts))\n",
    "        # -1 for no self count. percentage mutated. \n",
    "        perc = 1 - ((count_dict[ref_epitope]-1)/(np.sum(counts)-1))\n",
    "        \n",
    "        break\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MYSFVSEE', 'MYSFVSEE', 'MYSFVSEE', ..., 'MYSFVSEE', 'MYSFVSEE',\n",
       "       'MYSFVSEE'], dtype='<U8')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['M', 'Y', 'S', ..., 'S', 'E', 'E'],\n",
       "       ['M', 'Y', 'S', ..., 'S', 'E', 'E'],\n",
       "       ['M', 'Y', 'S', ..., 'S', 'E', 'E'],\n",
       "       ...,\n",
       "       ['M', 'Y', 'S', ..., 'S', 'E', 'E'],\n",
       "       ['M', 'Y', 'S', ..., 'S', 'E', 'E'],\n",
       "       ['M', 'Y', 'S', ..., 'S', 'E', 'E']], dtype='<U1')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epi_columns[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MYSFVLEE', 'MYSFVSDE', 'MYSFVSEE'], dtype='<U8')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([   1,    1, 5133])"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.005362633928857546"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MYSFVSEE'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_epitope"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MYSFVLEE': 1, 'MYSFVSDE': 1, 'MYSFVSEE': 5133}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.00038955979742894"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "perc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.005363</td>\n",
       "      <td>0.000390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.000584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index protein   epitope  start_pos  epi_len   entropy  perc_mutated\n",
       "0      0       E  MYSFVSEE          0        8  0.005363      0.000390\n",
       "1      1       E  YSFVSEET          1        8  0.008044      0.000584\n",
       "2      2       E  SFVSEETG          2        8  0.008044      0.000584\n",
       "3      3       E  FVSEETGT          3        8  0.008044      0.000584\n",
       "4      4       E  VSEETGTL          4        8  0.008044      0.000584"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data/processed/epitope_entropies.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11069"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df.entropy == 0.0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5826, 75)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ref_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(epi_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['E', 46, 25, 0.029753152247977425, 0.002403020940611067],\n",
       " ['E', 47, 25, 0.04479404697550162, 0.003604531410916545],\n",
       " ['E', 48, 25, 0.056493651187096074, 0.004634397528321288],\n",
       " ['E', 49, 25, 0.056493651187096074, 0.004634397528321288],\n",
       " ['E', 50, 25, 0.056493651187096074, 0.004634397528321288]]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_res[-5:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12.508289986140348"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.log2(len(seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(df['E'])[:,2];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([215., 436., 170.,  14.,  37.,   9.,   5.,  18.,  20.,  18.]),\n",
       " array([0.        , 0.00564937, 0.01129873, 0.0169481 , 0.02259746,\n",
       "        0.02824683, 0.03389619, 0.03954556, 0.04519492, 0.05084429,\n",
       "        0.05649365]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAN0UlEQVR4nO3dfayk5VnH8e9PtoDWyLZwaHCX9izpagATS90ASW1iwBdetLumkNCYsmlINo2YtKmNXVoTxRgD/iG1qWlDSuPWGIFSlU2pMYQXTU1KXcpbcYMsiGWFdLflpVIEg17+MTd42D1nz+yZmTNzbr6fZDLPcz/3zFxXZs/v3PvMy0lVIUnqy49MuwBJ0vgZ7pLUIcNdkjpkuEtShwx3SerQumkXAHDSSSfV/Pz8tMuQpDXl3nvv/V5VzS12bCbCfX5+nj179ky7DElaU5L8+1LHPC0jSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdmolPqK5V8ztvm8rjPnHNxVN5XElrhyt3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDg0d7kmOSXJfkq+2/U1J7knyaJKbkhzbxo9r+/va8fnJlC5JWsrRrNw/AuxdsH8tcF1VbQaeBa5o41cAz1bVO4Hr2jxJ0ioaKtyTbAQuBr7Q9gOcB9zSpuwCtrXtrW2fdvz8Nl+StEqGXbl/Gvgd4H/b/onAc1X1StvfD2xo2xuAJwHa8efb/NdJsiPJniR7Dh48uMLyJUmLWTbck/wqcKCq7l04vMjUGuLY/w9UXV9VW6pqy9zc3FDFSpKGM8yf2XsP8L4kFwHHAz/BYCW/Psm6tjrfCDzV5u8HTgX2J1kHnAA8M/bKJUlLWnblXlVXVdXGqpoHLgPurKrfAO4CLmnTtgO3tu3dbZ92/M6qOmzlLkmanFHe5/4J4GNJ9jE4p35DG78BOLGNfwzYOVqJkqSjNcxpmddU1d3A3W37ceDsRea8BFw6htokSSvkJ1QlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUoeWDfckxyf5ZpIHkjyc5Oo2vinJPUkeTXJTkmPb+HFtf187Pj/ZFiRJhxpm5f4ycF5V/SzwLuCCJOcC1wLXVdVm4Fngijb/CuDZqnoncF2bJ0laRcuGew280Hbf1C4FnAfc0sZ3Adva9ta2Tzt+fpKMrWJJ0rKGOuee5Jgk9wMHgNuBx4DnquqVNmU/sKFtbwCeBGjHnwdOXOQ+dyTZk2TPwYMHR+tCkvQ6Q4V7Vf1PVb0L2AicDZy+2LR2vdgqvQ4bqLq+qrZU1Za5ublh65UkDeGo3i1TVc8BdwPnAuuTrGuHNgJPte39wKkA7fgJwDPjKFaSNJxh3i0zl2R92/5R4BeBvcBdwCVt2nbg1ra9u+3Tjt9ZVYet3CVJk7Nu+SmcAuxKcgyDXwY3V9VXk/wLcGOSPwTuA25o828A/iLJPgYr9ssmULck6QiWDfeqehA4a5Hxxxmcfz90/CXg0rFUJ0laET+hKkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdWjdtAsY1fzO26ZdgiTNHFfuktQhw12SOmS4S1KHlg33JKcmuSvJ3iQPJ/lIG39rktuTPNqu39LGk+QzSfYleTDJuyfdhCTp9YZZub8C/HZVnQ6cC1yZ5AxgJ3BHVW0G7mj7ABcCm9tlB/C5sVctSTqiZcO9qp6uqm+17f8E9gIbgK3ArjZtF7CtbW8FvlQD3wDWJzll7JVLkpZ0VOfck8wDZwH3AG+rqqdh8AsAOLlN2wA8ueBm+9vYofe1I8meJHsOHjx49JVLkpY0dLgn+XHgK8BHq+oHR5q6yFgdNlB1fVVtqaotc3Nzw5YhSRrCUOGe5E0Mgv0vq+qv2/B3Xz3d0q4PtPH9wKkLbr4ReGo85UqShjHMu2UC3ADsrao/WXBoN7C9bW8Hbl0wfnl718y5wPOvnr6RJK2OYb5+4D3AB4GHktzfxj4JXAPcnOQK4DvApe3Y14CLgH3Ai8CHxlqxJGlZy4Z7VX2dxc+jA5y/yPwCrhyxLknSCPyEqiR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDa/4PZL8RTfOPgj9xzcVTe2xJw3PlLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUoWXDPckXkxxI8u0FY29NcnuSR9v1W9p4knwmyb4kDyZ59ySLlyQtbpiV+58DFxwythO4o6o2A3e0fYALgc3tsgP43HjKlCQdjWXDvar+EXjmkOGtwK62vQvYtmD8SzXwDWB9klPGVawkaTgrPef+tqp6GqBdn9zGNwBPLpi3v40dJsmOJHuS7Dl48OAKy5AkLWbcL6hmkbFabGJVXV9VW6pqy9zc3JjLkKQ3tpWG+3dfPd3Srg+08f3AqQvmbQSeWnl5kqSVWGm47wa2t+3twK0Lxi9v75o5F3j+1dM3kqTVs265CUn+CvgF4KQk+4HfA64Bbk5yBfAd4NI2/WvARcA+4EXgQxOoWZK0jGXDvao+sMSh8xeZW8CVoxYlSRqNn1CVpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1aN20C5CGMb/ztqk87hPXXDyVx5VG5cpdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QO+T53STPDzzOMj+Eu6XWmFbDTNM2eJ/WLxdMyktQhV+46Km/EVZ20Frlyl6QOGe6S1CFPy0hH0OMLbXpjmEi4J7kA+FPgGOALVXXNJB5H6pmvb2gUYz8tk+QY4M+AC4EzgA8kOWPcjyNJWtokzrmfDeyrqser6r+BG4GtE3gcSdISJnFaZgPw5IL9/cA5h05KsgPY0XZfSPLICh/vJOB7K7ztLLOvtcW+1paZ6SvXjnTzdyx1YBLhnkXG6rCBquuB60d+sGRPVW0Z9X5mjX2tLfa1tvTa10KTOC2zHzh1wf5G4KkJPI4kaQmTCPd/BjYn2ZTkWOAyYPcEHkeStISxn5apqleS/Bbw9wzeCvnFqnp43I+zwMindmaUfa0t9rW29NrXa1J12OlwSdIa59cPSFKHDHdJ6tBMh3uSC5I8kmRfkp2LHD8uyU3t+D1J5hccu6qNP5LkV1az7uWstK8kJya5K8kLST672nUvZ4S+finJvUkeatfnrXbtRzJCX2cnub9dHkjy66td+1JG+dlqx9/e/h1+fLVqHsYIz9V8kv9a8Hx9frVrH7uqmskLgxdjHwNOA44FHgDOOGTObwKfb9uXATe17TPa/OOATe1+jpl2T2Po683AzwMfBj477V7G2NdZwE+27Z8B/mPa/Yyprx8D1rXtU4ADr+6v1Z4WHP8K8GXg49PuZ0zP1Tzw7Wn3MM7LLK/ch/kag63ArrZ9C3B+krTxG6vq5ar6N2Bfu79ZsOK+quqHVfV14KXVK3doo/R1X1W9+lmIh4Hjkxy3KlUvb5S+XqyqV9r48SzyYb4pGeVniyTbgMcZPFezZKS+ejPL4b7Y1xhsWGpO+yF6HjhxyNtOyyh9zbJx9fV+4L6qenlCdR6tkfpKck6Sh4GHgA8vCPtpWnFPSd4MfAK4ehXqPFqj/hvclOS+JP+Q5L2TLnbSZvn73If5GoOl5gz1FQhTMkpfs2zkvpKcCVwL/PIY6xrVSH1V1T3AmUlOB3Yl+buqmvb/vEbp6Wrguqp6YQYXvKP09TTw9qr6fpKfA/42yZlV9YNxF7laZnnlPszXGLw2J8k64ATgmSFvOy2j9DXLRuoryUbgb4DLq+qxiVc7vLE8X1W1F/ghg9cUpm2Uns4B/jjJE8BHgU+2Dy3OghX31U7hfh+gqu5lcO7+pyZe8QTNcrgP8zUGu4HtbfsS4M4avDqyG7isvTK+CdgMfHOV6l7OKH3NshX3lWQ9cBtwVVX906pVPJxR+trUAoQk7wB+Gnhidco+ohX3VFXvrar5qpoHPg38UVXNyju3Rnmu5jL4WxQkOY1BZjy+SnVPxrRf0T3SBbgI+FcGv0U/1cb+AHhf2z6ewSv2+xiE92kLbvupdrtHgAun3csY+3qCwQrqBQarkDNWu/5x9wX8LoNV7f0LLidPu58x9PVBBi863g98C9g27V7G8W9wwX38PjP0bpkRn6v3t+fqgfZc/dq0exn14tcPSFKHZvm0jCRphQx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1KH/A+VCqV6gqkK9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(np.array(df['E'])[:,2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
